[Spec] TypeSpec to SDK Release Workflow - Updated Design (#16176)
Azure/azure-sdk-tools: tools/azsdk-cli/docs/specs/typespec-to-sdk-release-workflow.spec.md

|------|------|-------|
| Language CI pipelines | Build, test, lint, package validation | Language teams |
| SDK breaking change detector | Detects breaking changes in generated SDK API surface. Being combined into validation check. | @raych1 / @catalinaperalta |
| APIView (current) | SDK public API surface review via web UI | APIView team |
| **API Review Hub** (replacing APIView) | Creates synthetic review PRs with `API.md` diffs. PRs never merged — exist only for review. Architects auto-assigned. Approval recorded in ADO Package Work Items (API hash). CI gates release by checking hash. | @tjprescott |
| API review feedback resolution agent | Helps resolve API review comments via TypeSpec changes | azsdk-cli team |
| Pipeline troubleshooting agent | Diagnoses CI failures | azsdk-cli team |
| Auto SDK PR repair | `auto-sdk-build-fix` label → Copilot cloud agent fixes custom code drift → regenerate → rebuild. Shared orchestration in `eng/common/`, per-language opt-in. | azsdk-cli team |

#### Gaps

1. SDK breaking change detection integration in progress (being combined into validation check).
2. Auto-repair only handles custom-code drift — not all CI failure types.

> 3. ARH review PR creation on SDK PRs is not automated — mechanism TBD (open design gap).

Reviewer: Right now, this is triggered on-demand via the azsdk-cli/Azure MCP/release agent. However, it would not be difficult to trigger this automatically if we have the right trigger.

Author: Understood, documented the current state and since based on our discussion it seemed like the automation path was still to be figured out I left it as TBD.  Should that be ok?

4. API review feedback resolution agent needs evaluation for ARH compatibility.
5. Release gates transitioning from APIView → both → ARH only.

#### Open questions

- [ ] What triggers ARH review PR creation? SDK PR creation? Manual? Label?

> > **Note**: `<lang>-api-approved` labels are **informational** — source of truth is ADO Package Work Items (API hash). ARH will assign `<lang>-api-approved` labels on SDK PRs automatically when architect approves in future.

Reviewer: The source of truth is not the ADO work items (they are also, ironically, informational). It is the API hash stored in the ARH database (originally this was intended to be ADO--you aren't imaging things!). But yes, ARH will put the "api-approved" labels on the working PR that a review is linked to.

Author: Would it be fair to drop the "source of truth" wording and describe this as API hash stored truth or should I keep it as is?

| No state detection | Agent can't determine "where am I?" | Add workflow status tool (query release plan + PR status) |
| Errors not structured for agents | Some errors buried in logs | Every tool returns parseable errors with suggested next action |
| Label-driven automation gaps | Routing not fully connected | Connect label events to automation |

</details>

### Related process documentation

| Process | Link | Scope |
|---------|------|-------|
| Namespace approval | [Namespace approval (PR #44085)](https://github.com/Azure/azure-rest-api-specs/pull/44085) | Permissions, flow, labels — in progress |
| ARM review | [ARM review](https://eng.ms/docs/products/azure-developer-experience/design/api-specs-pr/arm-review) | ARM-specific gates |
| REST API spec review | [Review process](https://eng.ms/docs/products/azure-developer-experience/design/api-review) | Architect board flow |
| SDK API review (bridge) | [Arch board review process](https://github.com/Azure/azure-sdk/blob/main/.github/workflows/src/arch-board-review/ARCH-BOARD-REVIEW-PROCESS.md) | GitHub Form — **bridge** until ARH |

> | API Review Hub | TBD | Synthetic review PRs replacing APIView |

Reviewer: Very ephemeral as this POC transitions to a production implementation, but this has it all:
https://github.com/tjprescott/azure-sdk-tools/pull/49

| 3 | SDK PR not fully release-ready after generation: linter failures, test failures, merge conflicts, missing changelog/metadata. Tracked in [#15705](https://github.com/Azure/azure-sdk-tools/issues/15705). | 4 | @prkannap / Language teams | Yes | Open |
| 4 | Release trigger not automated — auto-trigger on SDK PR merge being built | 5 | @raych1 | Yes | In progress |
| 5 | ARH review PR creation not automated on SDK PRs | 4 | @tjprescott | Yes | Open |
| 6 | Breaking change findings require manual resolution | 1, 2 | @markcowl / @catalinaperalta | No | Open |
| 7 | `BreakingChangeReviewRequired` label routing to correct review team undefined | 2 | @raych1 / @markcowl / @catalinaperalta | No | Open |
| 8 | SDK breaking change detection integration in progress | 4 | @raych1 / @catalinaperalta | No | In progress |
| 9 | ARM vs data plane review gates not documented in detail (different review teams, labels, and blocking behavior) | 2, 4 | @samvaity / @prkannap | No | Open |
| 10 | Release pipeline provisioning delay for new RPs | 5 | EngSys | No | In progress |
| 11 | API review feedback agent needs ARH compatibility | 4 | azsdk-cli team | No | Open |

> | 12 | Release gate transition (APIView → ARH) | 5 | @tjprescott / EngSys | No | Open |

Reviewer: Met with Daniel and Ray to begin discussion on this. Essentially the existing pipelines will be updated to check BOTH APIView and ARH. If there is an approval in either system it will allow the release. Gradually (or suddenly!) APIView capability will be disabled, and the code paths removed.

>
> **Open question**: Should the no-spec-change and no-spec paths be documented as first-class entry points in this document? See also [#16297](https://github.com/Azure/azure-sdk-tools/issues/16297).

> 🧑‍💻 = requires human approval or manual gating

### The five-stage workflow

```mermaid
flowchart TD
    A["1. TypeSpec Authoring"] --> B["2. Spec PR Validation"]
    B --> C["3. SDK Generation (automatic)"]
    C --> D["4. SDK PR Validation & API Review"]
    D --> E["5. Release Coordination"]
```

### Service team journey

1. **Author TypeSpec**
   - Write `.tsp` files and `tspconfig.yaml` locally
   - Or use `azure-typespec-author` agent skill for guided authoring
   - Compile and lint locally before opening a PR

2. **Open a spec PR**
   - Push to `azure-rest-api-specs` and open a PR
   - CI validates automatically:
     - TypeSpec compilation (includes linting)
     - Suppression review (suppressed lint rules + suppressions.yaml changes)
     - Breaking change detection (spec-level: same-version + cross-version)
     - APIView token generation (for SDK-level review at Stage 4)
     - SDK generation dry-run (spec-gen-sdk)

>      - Labels applied based on results

Reviewer 2: Missing 'SDK breaking change detection' step that is already part of current mgmt-plane spec scenario.

Author: Added 'SDK breaking change detection (management plane only)' to the CI validation list. Thanks for catching that.

5. All approvals → PR merges

#### Failure paths

| Failure | Signal | Owner | Next action | Resolution |
|---------|--------|-------|-------------|------------|
| TypeSpec compile failure | CI red + compile error | Author | Fix TypeSpec | Manual |
| LintDiff violation | CI warning/error | Author | Fix or request suppression | Manual |
| Breaking change detected | `BreakingChangeReviewRequired` label | @raych1 / @markcowl / @catalinaperalta | Approve, suppress, or redesign | Manual |
| SDK dry-run failure | CI failure in spec-gen-sdk | Author / EngSys | Fix TypeSpec or escalate | Manual |
| Namespace needs approval | `namespace-<lang>-pending` label | Namespace approvers | Apply `namespace-<lang>-approved` | Manual (label) |

<details>
<summary>Deep spec: tools, contracts, and unresolved questions</summary>

#### Tool contract

| Tool | Role | Owner |
|------|------|-------|
| Spec PR validation pipeline | Orchestrates full validation suite | EngSys |
| TypeSpec compiler | CI compilation | TypeSpec team |
| TypeSpec Lintdiff | TypeSpec-native linting on PR diffs. Replacing Swagger-based Spectral LintDiff. Includes suppression process (@catalinaperalta). | EngSys / TypeSpec team |

> | `@azure-tools/typespec-breaking-change` | Phase A + B detection at **TypeSpec/spec level**. Auto-adds `BreakingChangeReviewRequired` / `VersioningReviewRequired` labels. Detects breaking changes in the API spec itself. | @markcowl |

Reviewer 3: ** Spec Level vs SDK-Level**  These are different concerns, but there will be some overlap:
- Deleted types will show up in both
- Narrowed types in request parameters is a breaking change in both, Widened types in request parameters is not an api breaking change but is likely an SDK breaking change
- Widened types in responses are a breaking change for both, narrowed response types do not break APIs, byt may break SDKs
- Added responses are a breaking api change, they *may be* an sdk breaking change
- Added optional parameters may break SDKs, but do not break APIs
- Parameter order changes break sdks but not apis
- Type name changes do not break APIs but break SDKs

Author: Thanks for the detail. Added a note to the SDK breaking change detector description clarifying the complementary nature: some changes break SDKs but not APIs and vice versa, so both detectors are needed for complete coverage.

| SDK breaking change detector ([PR #15588](https://github.com/Azure/azure-sdk-tools/pull/15588)) | Detects breaking changes in the **generated SDK API surface** (complements spec-level detection). Reports changes that may not be visible at spec level but affect SDK consumers. Being integrated into spec CI validation. | @raych1 / @catalinaperalta |
| APIView emitter (`typespec-apiview`) | Generates API surface tokens for SDK-level architect review (tokens used at Stage 4). **Will be retired with ARH.** | APIView team |
| spec-gen-sdk | SDK generation validation (dry-run) | EngSys (@prkannap) |
| Avocado / OAV | Legacy Swagger validation — **being deprecated** as TypeSpec-native tooling replaces them. | EngSys |

#### Gaps

1. Validation steps run independently — no designed chain for failure ordering or unified PR comment.
2. No endpoint liveness verification before spec PR merge — SDK may be generated for an undeployed API. *(Aspirational)*
3. Avocado/OAV deprecation still in progress.

> 4. `BreakingChangeReviewRequired` label routing to correct review team is undefined (CODEOWNERS, custom Action, or DevOps?).

Reviewer 3: VersioningReviewRequired is the label for phase A
BreakingChangeReviewRequired is the label for phase B

We need to decide whether to have a different label for SDK breaking changes, or if BreakingChangeReviewRequired covers both

Author: Resolved by Ray's commit (b49e13e): SDK breaking changes use separate `BreakingChange-{Language}-Sdk` labels, distinct from `BreakingChangeReviewRequired` (spec-level). Also added `VersioningReviewRequired` = Phase A to the spec.

#### Open questions

- [ ] How should `BreakingChangeReviewRequired` label route to the correct review team? CODEOWNERS, custom Action, or DevOps?
- [ ] Should spec-gen-sdk failures be PR comments, structured JSON, or both?

> - [ ] How do SDK breaking change detector findings (PR #15588) differ from `@azure-tools/typespec-breaking-change` findings? What labels does the SDK detector add, and how are its errors reported?

Reviewer 3: Findings are surfaced as diiagnostics,  the diagnostic type and info should allow automation to figure out the details of a breaking change (using target and diagnostic type).  

If we need additional data, we could output json detail for each breaking change type.

Author: Noted. The diagnostic type + target info should be sufficient for agent-assisted resolution. If structured JSON output is available, we can reference it in the tool contract for automation purposes.

     - SDK generation dry-run (spec-gen-sdk)
     - Labels applied based on results

3. 🧑‍💻 **Wait for approvals** *(human gating)*
   - **`PublishToCustomers` label** — required on all PRs targeting `main` or `RPSaaSMaster`. Author self-applies to acknowledge APIs are shipped to customers. Enforced by CI (`summarize-checks`). **Candidate for removal** -- redundant once other gates pass; causes service team confusion.
   - **Namespace approval** — required for first preview of new packages (`namespace-<lang>-pending` → `namespace-<lang>-approved`)
   - **ARM review** — required for ARM/management-plane specs (`ARM-Review-Required` → `ARMSignedOff`)
   - **Breaking change review** — Spec-level: `BreakingChangeReviewRequired` label auto-applied by CI on any spec PR with breaking changes (review team defined for ARM, undefined for data-plane -- Gap #9). SDK-level: `BreakingChange-{Language}-Sdk` label currently applies to management plane spec PRs only.
   - > **Note**: There is no spec-level API review. API review happens only at the SDK level (Stage 4).

4. **Spec PR merges → SDK generation is automatic**
   - Release plan work item created

>    - SDK code generated per language via emitters

Reviewer 4: ```suggestion
   - SDK code generated per language via emitters. SDK Pull request is generated automatically for management plane only. Service team needs to generate the SDK manually using azsdk agent for data plane.
```

Author: Applied. Updated the quick-start and Stage 3 to distinguish management plane (automatic) vs data-plane (manual via azsdk agent).

   - Customizations applied (`client.tsp` + code customizations)
   - SDK PRs opened in each language repo and linked to release plan

>    - ⚠️ Generation failures currently fail silently — [Gap #4](#gap-tracker)

Reviewer 4: SDK generation failures are visible on the release plan dashboard.

Author: Updated Gap #2 and Stage 3 to note that failures are visible on the release plan dashboard. Kept the gap open for structured diagnostics (agent-actionable reporting).

5. 🧑‍💻 **SDK PR review & approval** *(human gating)*
   - SDK CI runs automatically:
     - Build → test → lint → package validation
     - SDK breaking change detection
     - APIView generates SDK API surface review (current) or API Review Hub creates review PR (future)
   - **API review**: Architects review generated SDK public API surface
     - `<lang>-api-approved` labels (current) are **informational** — source of truth is the ARH database (API hash)
     - ARH will assign `api-approved` label on SDK PRs automatically when architect approves (replacing per-language labels)
   - **Auto-repair**: `auto-sdk-build-fix` label triggers Copilot agent to fix custom code drift
   - **ARM SDK PRs**: Reviewed by Haoling/Shanghai team with release plans attached

6. 🧑‍💻 **Release** *(human gating)*

>    - Release pipeline triggered (becoming automatic — @raych1)

Reviewer 4: This has already been enabled.

Author: Marked as done in the gap tracker (Gap #4) and updated the Stage 5 text.

>    - Release pipeline triggered (becoming automatic — @raych1)

Reviewer 4: ```suggestion
   - The release pipeline is automatically triggered when an SDK pull request is merged, provided the auto-release label is applied to the pull request. (becoming automatic — @raych1)
```

Author: Applied. Updated throughout to reflect auto-release label trigger.

   - Packages published → release plan completes → Service Tree KPI updated

### For Reviewers

1. **ARM review** (ARM specs only) — Review resource model correctness on spec PRs; apply `ARMSignedOff` label
2. **SDK API review** — Review generated SDK API surface on SDK PRs via APIView (current) or API Review Hub review PRs (future). **There is no spec-level API review** — API review applies only to the generated SDK. **Current**: `<lang>-api-approved` labels applied manually by architects. **Future (ARH)**: `api-approved` label applied automatically via webhooks/GH App. Labels are **informational** — source of truth is the ARH database (API hash).
3. **SDK PR review** (Haoling/Shanghai team, management plane only) — Review generated SDK PRs that have a release plan attached; approve & merge
4. **Breaking change review** — Spec-level: `BreakingChangeReviewRequired` label on spec PRs (review team defined for ARM; data-plane routing is Gap #9). SDK-level: `BreakingChange-{Language}-Sdk` label on management plane spec PRs only.
5. **Namespace review** — Approve new package namespaces; apply `namespace-<lang>-approved` labels

> 6. **Release approval** — Approve release pipeline runs (Haoling/Shanghai team for ARM)

Reviewer 4: ```suggestion
6. **Release approval** — Approve release pipeline runs (Shanghai SDK team for ARM)
```

Author: Applied, updated to Shanghai SDK team for ARM throughout.

### For EngSys / SDK Team

1. **Monitor pipelines** — Spec PR validation, SDK generation, and SDK CI pipelines run automatically
2. **Label routing** — Labels like `BreakingChangeReviewRequired`, `namespace-<lang>-pending`, and `auto-sdk-build-fix` trigger review routing and automation
3. **Generation failures** — When SDK generation fails, diagnose via pipeline logs (structured error reporting is a [known gap](#gap-tracker))
4. **Auto-repair** — `auto-sdk-build-fix` label triggers Copilot cloud agent to fix custom code drift on SDK PRs
5. **Release coordination** — `azsdk_release_sdk` checks readiness; release pipelines publish to package registries (manual approval gate required)
6. **Track progress** — [Release plan dashboard](https://aka.ms/azsdk/releaseplan-dashboard) shows where each service is in the process

> 

Reviewer 4: We should add GitHub.io docs automation here that creates a PR in azure-sdk repo to update the package index data visibile in aka.ms/azsdk(Azure SDK GitHub io docs). PR needs to be reviewed, approved and merged by each language team.

Author: Added GitHub.io docs automation as a step in the release section and Stage 5 happy path.

#### Happy path

1. Spec PR merges → pipeline triggers automatically
2. **Stage A**: Create or find release plan work item
3. **Stage B**: For each language: tsp-client syncs → emitter generates → customizations applied → build → test → metadata updated → SDK PR created and linked to release plan

#### Failure paths

| Failure | Signal | Owner | Next action | Resolution |
|---------|--------|-------|-------------|------------|
| Generation failure (any language) | Failed pipeline check (buried in logs) | EngSys / language owner | Investigate logs manually | Manual |
| Customization drift | Build failure in SDK PR | azsdk-cli team | `auto-sdk-build-fix` label → auto-repair | Automatic |
| Release plan creation failure | Pipeline failure | azsdk-cli team | Investigate DevOps connectivity | Manual |

<details>
<summary>Deep spec: tools, contracts, and unresolved questions</summary>

#### Tool contract

| Tool | Role | Owner |
|------|------|-------|
| tsp-client | Syncs TypeSpec project into SDK repo | @catalinaperalta (tool), EngSys (language repo integration) |
| Language emitters | Generate client library code (one per language) | Language teams |

> | spec-gen-sdk | Pipeline automation — runs full workflow, creates SDK PRs | EngSys (@prkannap) |

Reviewer 4: Incorrect owner tag

Author: Fixed, updated to @raych1.

   - Write `.tsp` files and `tspconfig.yaml` locally
   - Or use `azure-typespec-author` agent skill for guided authoring
   - Compile and lint locally before opening a PR

2. **Open a spec PR**
   - Push to `azure-rest-api-specs` and open a PR
   - CI validates automatically:
     - TypeSpec compilation (includes linting)
     - Suppression review (suppressed lint rules + suppressions.yaml changes)
     - Breaking change detection (spec-level: same-version + cross-version)
     - SDK breaking change detection (management plane only, detects SDK API surface breaks)
     - APIView token generation (for SDK-level review at Stage 4)
     - SDK generation dry-run (spec-gen-sdk)
     - Labels applied based on results

3. 🧑‍💻 **Wait for approvals** *(human gating)*
   - **`PublishToCustomers` label** — required on all PRs targeting `main` or `RPSaaSMaster`. Author self-applies to acknowledge APIs are shipped to customers. Enforced by CI (`summarize-checks`). **Candidate for removal** -- redundant once other gates pass; causes service team confusion.
   - **Namespace approval** — required for first preview of new packages (`namespace-<lang>-pending` → `namespace-<lang>-approved`)

>    - **ARM review** — required for ARM/management-plane specs (`ARM-Review-Required` → `ARMSignedOff`)

Reviewer 5: They have now a new step called ARM leaving modeling review, I can give you more details. Long story short, we don't own this, so it's fine if we're not specific.

Maybe we should anticipate that ARM has control to add whatever checks they want, so phrsing it like:

- ARM Validation (owned by ARM). Example (not exhaustive)
  - ARM review board
  - Breaking change review
  - Modeling requiring (leasing) review

   - **Breaking change review** — Spec-level: `BreakingChangeReviewRequired` label auto-applied by CI on any spec PR with breaking changes (review team defined for ARM, undefined for data-plane -- Gap #9). SDK-level: `BreakingChange-{Language}-Sdk` label currently applies to management plane spec PRs only.
   - > **Note**: There is no spec-level API review. API review happens only at the SDK level (Stage 4).

> 4. **Spec PR merges → SDK generation is automatic (management plane)**

Reviewer 5: As we discussed, maybe it's more "convenience or not convenience". Just a  note, not blocking

Author: Will note it as aspirational want for "convenience or not convenience" but keeping it true to what it exists as of today.

   - The SDK PR is unique per release plan -- subsequent spec PRs (e.g., client.tsp customizations) regenerate into the same SDK PR
   - Customizations applied (`client.tsp` + code customizations)
   - SDK PRs opened in each language repo and linked to release plan
   - Generation failures visible on [release plan dashboard](https://aka.ms/azsdk/releaseplan-dashboard)

5. 🧑‍💻 **SDK PR review & approval** *(human gating)*
   - SDK CI runs automatically:
     - Build → test → lint → package validation
     - SDK breaking change detection
     - APIView generates SDK API surface review (current) or API Review Hub creates review PR (future)
   - **API review**: Architects review generated SDK public API surface
     - `<lang>-api-approved` labels (current) are **informational** — source of truth is the ARH database (API hash)
     - ARH will assign `api-approved` label on SDK PRs automatically when architect approves (replacing per-language labels)
   - **Auto-repair**: `auto-sdk-build-fix` label triggers Copilot agent to fix custom code drift
   - **ARM SDK PRs**: Reviewed by Shanghai SDK team with release plans attached

6. 🧑‍💻 **Release** *(human gating)*

>    - The release pipeline is automatically triggered when an SDK pull request is merged, provided the `auto-release` label is applied to the pull request.

Reviewer 5: We should write somewhere that this label is auto-applied for ARM

>    - Manual approval gate required for security (cannot be removed; ARM approval = Shanghai SDK team)

Reviewer 5: This changed after my last discussion with Mayuri, now this is service team

   - Release gate: API Review Hub verifies approved API hash
   - Packages published → release plan completes → Service Tree KPI updated

>    - GitHub.io docs automation creates PR in azure-sdk repo to update package index (aka.ms/azsdk). Reviewed and merged by each language team.

Reviewer 5: We should automate that, just a note that for me it's a gap that it's reviewed

### For Reviewers

1. **ARM review** (ARM specs only) — Review resource model correctness on spec PRs; apply `ARMSignedOff` label
2. **SDK API review** — Review generated SDK API surface on SDK PRs via APIView (current) or API Review Hub review PRs (future). **There is no spec-level API review** — API review applies only to the generated SDK. **Current**: `<lang>-api-approved` labels applied manually by architects. **Future (ARH)**: `api-approved` label applied automatically via webhooks/GH App. Labels are **informational** — source of truth is the ARH database (API hash).

> 3. **SDK PR review** (Shanghai SDK team, management plane only) — Review generated SDK PRs that have a release plan attached; approve & merge

Reviewer 5: I believe today Shanghai is asking the service team to approve those PR as well, to be verified

4. **Breaking change review** — Spec-level: `BreakingChangeReviewRequired` label on spec PRs (review team defined for ARM; data-plane routing is Gap #9). SDK-level: `BreakingChange-{Language}-Sdk` label on management plane spec PRs only.
5. **Namespace review** — Approve new package namespaces; apply `namespace-<lang>-approved` labels

> 6. **Release approval** — Approve release pipeline runs (Shanghai SDK team for ARM)

Reviewer 5: As said before, should be service team for ARM

```

**Key insight**: After spec PR merge, the flow is largely **automated**. Service team re-engages only if SDK CI fails, API review has feedback, or release needs manual approval.

> **ARM vs Data Plane divergence** — Same high-level flow but diverge at review gates. ARM requires ARM review sign-off + stricter resource model constraints. Data plane skips ARM review.
>
> **Open question: Data-plane review model** — With the stewardship board being reconsidered, who reviews data-plane PRs? An emerging pattern: for ARM, if the REST API spec looks fine, we assume the SDK is ok to ship. For data-plane, if the SDK looks fine, we could assume the TypeSpec spec is ok to merge. This would mean ARM quality flows spec → SDK, while data-plane quality flows SDK → spec.

---

## Stage modules

<a name="stage-1-typespec-authoring"></a>
### Stage 1 — TypeSpec Authoring

#### Quick card

| Field | Value |
|-------|-------|
| **Purpose** | Author/update TypeSpec locally, validate, prepare for spec PR |
| **Entry signal** | Service team has API requirements |
| **Exit signal** | `.tsp` files compile, lint passes, ready for PR |

> | **Owners** | TypeSpec team (compiler/linter), Haoling/Shanghai (authoring agent), @markcowl (breaking change) |

Reviewer 5: Let's group breaking change with TypeSpec team, not Mark individually

- [ ] For patch releases — what triggers the workflow differently?

</details>

---

## Cross-cutting contracts

### Label contract

#### Spec PR Labels (`azure-rest-api-specs`)

| Label | Applied by | Meaning | Blocking? | Automation |
|-------|-----------|---------|-----------|------------|
| `BreakingChangeReviewRequired` | CI | Breaking change detected | Yes | ⚠️ Label auto, routing manual |
| `VersioningReviewRequired` | CI | Versioning review needed | Yes | ⚠️ Label auto, assignment manual |
| `ARM-Review-Required` | CI | ARM spec — routes to ARM team | Yes | ✅ Fully automated |
| `ARMSignedOff` | ARM team | ARM review approved | Unblocks | ✅ Manual label, gate automated |
| `APIStewardshipBoard-SignedOff` | Stewardship board | Data-plane REST API spec approved (stewardship review) | No (transitioning) | ⚠️ Process in transition |
| `namespace-<lang>-pending` | CI | New namespace detected | Yes | ✅ Fully automated |
| `namespace-<lang>-approved` | Architect | Namespace approved | Unblocks | ✅ Manual label, gate automated |
| `namespace-approved-all` | Architect | Approves all languages (mgmt) | Unblocks | ✅ Manual label, gate automated |

> | `Approved-BreakingChange` | Review team | Breaking change approved | Unblocks | ⚠️ Manual label, gate works |

Reviewer 5: Nit, there is not one, but a lot of breaking change approved labels

| `BreakingChange-{Language}-Sdk` | CI | Sdk breaking change detected (management plane only) | Yes | ✅ Fully automated |
| `BreakingChange-{Language}-Sdk-Approved` | Review team | Sdk breaking change approved | Yes | ⚠️ Manual label, validation automated |
| `BreakingChange-{Language}-Sdk-Suppression` | Authors | SDK breaking change suppression updates | Yes | ✅ Fully automated |
| `BreakingChange-{Language}-Sdk-Suppression-Approved` | Review team | SDK breaking change suppression approved | Yes | ⚠️ Manual label, validation automated |
| `PublishToCustomers` | Author | Acknowledges APIs are shipped to customers. Required for PRs targeting `main` or `RPSaaSMaster`. Without it, PR cannot merge. **Low-hanging fruit for removal** -- redundant once all other gates pass; causes confusion for service teams. Could be auto-applied or removed entirely. | Yes | ⚠️ Manual label, enforced by CI |

> #### SDK PR Labels (language repos)

Reviewer 5: missing auto-release

>
> **Open question**: Should the no-spec-change and no-spec paths be documented as first-class entry points in this document? See also [#16297](https://github.com/Azure/azure-sdk-tools/issues/16297).

> 🧑‍💻 = requires human approval or manual gating

### The five-stage workflow

```mermaid
flowchart TD
    A["1. TypeSpec Authoring"] --> B["2. Spec PR Validation"]
    B --> C["3. SDK Generation (automatic)"]
    C --> D["4. SDK PR Validation & API Review"]
    D --> E["5. Release Coordination"]
```

### Service team journey

1. **Author TypeSpec**
   - Write `.tsp` files and `tspconfig.yaml` locally
   - Or use `azure-typespec-author` agent skill for guided authoring
   - Compile and lint locally before opening a PR

2. **Open a spec PR**
   - Push to `azure-rest-api-specs` and open a PR
   - CI validates automatically:
     - TypeSpec compilation
     - TypeSpec LintDiff (guideline compliance)
     - Breaking change detection (Phase A + B)
     - APIView token generation (for SDK-level review at Stage 4)
     - SDK generation dry-run (spec-gen-sdk)

>      - Labels applied based on results

Reviewer 6: Not sure if we need this level of specificity but their tspconfig.yaml needs to have the typespec-autorest emitter configured in tspconfig.yaml and the generated swagger must be part of the PR

Reviewer 6: Also, will there be any requirements on folks keeping a README and updating it?

Author: For the readme, I was checking, is the readme.md still required for TypeSpec projects, or is tspconfig.yaml sufficient?

Reviewer 6: Today, README checks are still required on PRs, I think it's only done through the swagger checks, however this depends on what the plan is for continuing to check in typespec-generated swaggers. I believe if we keep checking in the swaggers that the readme will still be required. For example, it's part of the typespec project structure docs: https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/directory-structure.md

Reviewer 6: ^ that doc says the readme is required for tooling purposes

3. 🧑‍💻 **Wait for approvals** *(human gating)*

>    - **Namespace approval** — required for first preview of new packages (`namespace-<lang>-pending` → `namespace-<lang>-approved`)

Reviewer 6: Related to my comment above, this can only happen if the emitters are configured and checked in with the spec PR to the repo.

Author: Correct -- the spec assumes tspconfig.yaml with emitters is checked in. Stage 1 documents this as a prerequisite.

   - **ARM review** — required for ARM/management-plane specs (`ARM-Review-Required` → `ARMSignedOff`)
   - **Breaking change review** — required for ARM specs with breaking changes (`BreakingChangeReviewRequired` → `Approved-BreakingChange`)

>    - > **Note**: There is no spec-level API review. API review happens only at the SDK level (Stage 4).

Reviewer 6: How will data-plane breaking changes be reviewed/handled?

Author: My understanding is that this is currently done at the SDK generation level for data-plane.

>2. **SDK API review** — Review generated SDK API surface on SDK PRs via APIView (current) or API Review Hub review PRs (future). **There is no spec-level API review** — API review applies only to the generated SDK. **Current**: `<lang>-api-approved` labels applied manually by architects. **Future (ARH)**: `api-approved` label applied automatically via webhooks/GH App. Labels are **informational** — source of truth is the ARH database (API hash).

Author: Ray's commit (b49e13e) adds SDK-level breaking change detection with `BreakingChange-{Language}-Sdk` labels that cover both ARM and data-plane. The spec-level breaking change detection (`BreakingChangeReviewRequired`) and its data-plane scope will be documented as part of Gap # 9.

> 4. **Spec PR merges → SDK generation is automatic**

Reviewer 6: Today for spec PRs after all of the appropriate approvals service partners have to add a `PublishToCustomers` label to their PR, will we continue requiring that label as part of the process?

Author: I wasn’t aware of this label. As far as I know, we haven’t changed anything related to the `PublishToCustomers` label. That said, I’d like to understand what this label is for and whether there’s any process documentation for it.

Author: Documented. Added PublishToCustomers to the label contract table with its enforcement details. Can revisit later if the process needs simplification.

   - SDK CI runs automatically:
     - Build → test → lint → package validation
     - SDK breaking change detection
     - APIView generates SDK API surface review (current) or API Review Hub creates review PR (future)
   - **API review**: Architects review generated SDK public API surface
     - `<lang>-api-approved` labels (current) are **informational** — source of truth is the ARH database (API hash)
     - ARH will assign `api-approved` label on SDK PRs automatically when architect approves (replacing per-language labels)
   - **Auto-repair**: `auto-sdk-build-fix` label triggers Copilot agent to fix custom code drift
   - **ARM SDK PRs**: Reviewed by Haoling/Shanghai team with release plans attached

6. 🧑‍💻 **Release** *(human gating)*
   - Release pipeline triggered (becoming automatic — @raych1)
   - Manual approval gate required for security (cannot be removed; ARM approval = Haoling/Shanghai team)
   - Release gate: API Review Hub verifies approved API hash
   - Packages published → release plan completes → Service Tree KPI updated

### For Reviewers

1. **ARM review** (ARM specs only) — Review resource model correctness on spec PRs; apply `ARMSignedOff` label

> 2. **SDK API review** — Review generated SDK API surface on SDK PRs via APIView (current) or API Review Hub review PRs (future). **There is no spec-level API review** — API review applies only to the generated SDK. **Current**: `<lang>-api-approved` labels applied manually by architects. **Future (ARH)**: `api-approved` label applied automatically via webhooks/GH App. Labels are **informational** — source of truth is the ARH database (API hash).

Reviewer 6: Are there any concerns about having to go back and implement rest api level fixes when azure patterns arent implemented correctly for preview services, for example lro/paging patterns that wont work well in the sdks (when the typespec author skill isnt used or the default recommendation is not followed)? I'm just wondering if the sdk level review is coming too late if the specs repo pr is already merged in

Author: The SDK breaking change detector (@chunyu3/@raych1) runs at Stage 2 (spec PR CI), so SDK-impacting patterns like incorrect LRO/paging would be flagged before the spec merges -- not after. That is the direct mitigation for this concern. Gap # 9 tracks the broader question of documenting review gates across ARM and data-plane.

3. **SDK PR review** (Haoling/Shanghai team, management plane only) — Review generated SDK PRs that have a release plan attached; approve & merge

> 4. **Breaking change review** — Review breaking changes flagged by `BreakingChangeReviewRequired` label on spec PRs (ARM specs only)

Reviewer 6: what's the plan for data-plane breaking changes?

Author: Covered by Ray's commit (b49e13e): SDK-level breaking change detection now applies to both ARM and data-plane via `BreakingChange-{Language}-Sdk` labels. Spec-level data-plane breaking change handling is being documented as part of Gap #9.

     │                 │                │                │                │  SDK PR merged │
     │                 │                │                │                │                │
     │  Trigger release│                │                │                │       ┌───────>│
     │──────────────────────────────────────────────────────────────────────────┘        │
     │                 │                │                │                │   Publish pkgs │
     │  ✅ Done!       │                │                │                │   Update KPI   │
     │<────────────────────────────────────────────────────────────────────────────────────│
```

**Key insight**: After spec PR merge, the flow is largely **automated**. Service team re-engages only if SDK CI fails, API review has feedback, or release needs manual approval.

> **ARM vs Data Plane divergence** — Same high-level flow but diverge at review gates. ARM requires ARM review sign-off + stricter resource model constraints. Data plane skips ARM review. See: [ARM review](https://eng.ms/docs/products/azure-developer-experience/design/api-specs-pr/arm-review)
>

> > **Open question: Data-plane review model** — With the stewardship board being reconsidered, who reviews data-plane PRs? An emerging pattern: for ARM, if the REST API spec looks fine, we assume the SDK is ok to ship. For data-plane, if the SDK looks fine, we could assume the TypeSpec spec is ok to merge. This would mean ARM quality flows spec → SDK, while data-plane quality flows SDK → spec.

Reviewer 6: ok this is related to my question above about when to have the sdk review for data-plane prs, glad it's tracked here. Since we're trying to reduce back and forth it might be worthwhile to consider having the sdk review earlier in the process for data-plane specs

Author: Agreed on reducing back-and-forth. The SDK breaking change detector (@chunyu3/@raych1) runs at Stage 2 (spec PR CI), so SDK-impacting patterns are caught before the spec merges -- effectively giving earlier SDK-level signal for data-plane specs too. This is tracked in Gap # 9 and we have updated the spec to reflect both spec-level and SDK-level breaking change coverage.

|---------|--------|-------|-------------|------------|
| Compile error | TypeSpec compiler error | Author | Fix `.tsp` syntax/types | Manual |
| Linter violation | Linter warning/error | Author | Fix or suppress with decorator | Manual |
| Breaking change detected | Tool report with DiffKind + source location | Author | Apply suppression decorator or redesign | Manual |

<details>
<summary>Deep spec: tools, contracts, and unresolved questions</summary>

#### Tool contract

| Tool | Role | Owner |
|------|------|-------|
| TypeSpec compiler | Compile `.tsp` files, catch syntax/type errors | TypeSpec team |
| TypeSpec linter | Static guideline compliance (distinct from LintDiff) | TypeSpec team |
| TypeSpec authoring agent (`azure-typespec-author` skill) | Assist with ARM/data-plane patterns, Azure REST API guidelines | Haoling/Shanghai |
| `@azure-tools/typespec-breaking-change` | Phase A: same-version regression. Phase B: cross-version evolution. Inline suppression via decorators. | @markcowl |

#### Gap

Breaking change tool reports findings with DiffKind, source location, and suggested suppression decorator — but resolution is manual. No agent auto-resolves.

#### Next step

> Build author-validation loop where agent auto-applies suppression decorators based on structured breaking change output.

Reviewer 6: why is the next step focused on a suppression decorator instead of fixing the breaking change?

Reviewer 6: Are the breaking changes reported going to recommend how to fix the given breaking change?

Author: Agreed. Ray's commit (b49e13e) updated the failure path to 'Address breaking change or add suppression', fixing is listed first, suppression as fallback. The spec language is now consistent with this framing throughout.

Author: @chunyu3: confirming the SDK breaking change detector reports the breaking change with enough context for the author to fix. Please add any details about fix recommendations to the spec if applicable.

#### Open questions

- [ ] Does `@azure-tools/typespec-breaking-change` output provide enough structured context for agent auto-resolution? Need confirmation from @markcowl.

</details>

---

<a name="stage-2-spec-pr-validation"></a>
### Stage 2 — Spec PR Validation

#### Quick card

| Field | Value |
|-------|-------|
| **Purpose** | Validate spec PR: compile, lint, breaking changes, APIView tokens, SDK dry-run |
| **Entry signal** | PR opened/updated in `azure-rest-api-specs` |
| **Exit signal** | CI passes + review labels applied + approved & merged |

> | **Owners** | EngSys (pipeline), @raych1 / @catalinaperalta (breaking change), TypeSpec team (lintdiff) |

Reviewer 6: Nit: Mark owns the breaking change tool, I'll own the suppression review tool
```suggestion
| **Owners** | EngSys (pipeline), @raych1 / @markcowl (breaking change), TypeSpec team (lintdiff) |
```

Author: Fixed. @markcowl owns breaking change tool, @catalinaperalta owns suppression review tool. Updated throughout.

| **Reviewer ask** | Confirm CI ordering, label semantics, and blocking vs informational |

#### Happy path

1. PR opens → CI triggers automatically

> 2. TypeSpec compiles → LintDiff runs → breaking change detection → APIView tokens generated → SDK dry-run → SDK APIView generated

Reviewer 6: should we include something about the generated swagger? It's used by the downstream docs pipeline and currently still required with all Typespec projects

Author: Is the generated swagger requirement going away when docs pipeline moves to TypeSpec-native consumption, or is it long-term?

Reviewer 6: I believe the docs generation pipeline isnt going to change any time soon, but I dont have any insight onto whether there are any plans in the works to move that to TypeSpec

Author: Got it. Noted in the spec: generated swagger required for foreseeable future (docs pipeline dependency). Exception mechanism being designed for specs that don't emit swagger. Will update when that lands.

3. Labels applied based on results
4. ARM review (if ARM spec) + namespace approval (if new package)

> 5. All approvals → PR merges

Reviewer 6: This is where the author needs to add tha PublishToCustomers label or we can remove that label requirement if it's no longer necessary (maybe this diverges for ARM vs data-plane prs)

Author: @raych1 @catalinaperalta: the `PublishToCustomers` label is not currently in the spec's label contract. If it is still required for spec PR merge, please add it with the correct scope (ARM only, data-plane only, or both). If it is being retired, let me know so we can document the transition.

Author: Added to the spec. PublishToCustomers is now in the label contract as a required label for PRs targeting main/RPSaaSMaster. Can revisit if the process needs simplification.

#### Failure paths

| Failure | Signal | Owner | Next action | Resolution |
|---------|--------|-------|-------------|------------|
| TypeSpec compile failure | CI red + compile error | Author | Fix TypeSpec | Manual |

> | LintDiff violation | CI warning/error | Author | Fix or request suppression | Manual |

Reviewer 6: technically this will be a Suppression violation, the lint rules are run during the compile step and the suppression review tool reports suppressions on a specific set of rules (which will include all of the equivalent lint diff rules)

Author: Updated: renamed to 'Suppression violation' in failure paths and clarified that lint rules run during compile. Suppression review is the distinct step.

| Breaking change detected | `BreakingChangeReviewRequired` label | @raych1 / @markcowl / @catalinaperalta | Approve, suppress, or redesign | Manual |
| SDK dry-run failure | CI failure in spec-gen-sdk | Author / EngSys | Fix TypeSpec or escalate | Manual |
| SDK breaking change detected | `BreakingChange-{Language}-Sdk` label | Author | Address breaking change or add suppression | Manual |
| Namespace needs approval | `namespace-<lang>-pending` label | Namespace approvers | Apply `namespace-<lang>-approved` | Manual (label) |

<details>
<summary>Deep spec: tools, contracts, and unresolved questions</summary>

#### Tool contract

| Tool | Role | Owner |
|------|------|-------|
| Spec PR validation pipeline | Orchestrates full validation suite | EngSys |
| TypeSpec compiler | CI compilation | TypeSpec team |

> | TypeSpec Lintdiff | TypeSpec-native linting on PR diffs. Replacing Swagger-based Spectral LintDiff. Includes suppression process (@catalinaperalta). | EngSys / TypeSpec team |

Reviewer 6: the "TypeSpec-native linting" occurs during the typespec compile step and the next step would be to review suppressions for lint rules that ideally shouldnt be suppressed. We recently had a discussion on the right terminology so I think we should update this to refer to the Suppression Review process
```suggestion
| TypeSpec compiler | CI compilation, TypeSpec-native linting. | TypeSpec team |
| TypeSpec Suppression Review |  Review TypeSpec lint rules that are suppressed (Replacing Swagger-based Spectral LintDiff) (@catalinaperalta). | EngSys / TypeSpec team |
```

Reviewer 6: Also, there are additional suppressions to review in a suppression.yaml file, we could probably include it in this step although that review is manual. I do believe CI reports when additions/changes were made to the suppressions.yaml file, but I'm not sure which check reports it. It adds a `SuppressionReviewRequired` label to the PR and reviewers need to add a suppressions approved label. This is an important part of the process. Not sure if we should group it in this entry.

Reviewer 6: Also the new tool might have a new label, this is still pending

Author: Thanks for the terminology correction. Updated the spec: renamed to 'TypeSpec Suppression Review', changed 'LintDiff violation' to 'Suppression violation' in failure paths, and added suppressions.yaml review step. Please add the Suppression Review process details (review steps, label names, blocking behavior) directly to the branch, or link to existing documentation so I can incorporate it.

| `@azure-tools/typespec-breaking-change` | Phase A + B detection at **TypeSpec/spec level**. Auto-adds `BreakingChangeReviewRequired` / `VersioningReviewRequired` labels. Detects breaking changes in the API spec itself. | @markcowl |

> | SDK breaking change detector ([PR #15588](https://github.com/Azure/azure-sdk-tools/pull/15588)) | Detects breaking changes in the **generated SDK API surface** (complements spec-level detection). Reports changes that may not be visible at spec level but affect SDK consumers. Being integrated into spec CI validation. | @raych1 / @catalinaperalta |

Reviewer 6: Just a nit here I dont own the breaking changes tool, or is this referring to who will review breaking changes?

Author: Fixed. Removed you from breaking change ownership rows. You are listed only on suppression review tool and tsp-client.

| APIView emitter (`typespec-apiview`) | Generates API surface tokens for SDK-level architect review (tokens used at Stage 4). **Will be retired with ARH.** | APIView team |
| spec-gen-sdk | SDK generation validation (dry-run) | EngSys (@raych1) |

> | Avocado / OAV | Legacy Swagger validation — **being deprecated** as TypeSpec-native tooling replaces them. | EngSys |

Reviewer 6: I was wondering what we're doing in regards to these checks in the meantime, I know they will be deprecated, but at the moment if PRs need to be merged in the issues in this checks need to be overridden, addressed, or suppressed with a label in some instances

Author: Do you have a list of labels or suppressions or process guidance for this that can be included here. It does seem temporary but dont think we know it for sure and it would be good to document?

Reviewer 6: I was mostly asking to see if the other folks on the team had an opinion on this. For now, I think we're mostly overriding any issues that arent worth addressing

Author: Makes sense. The spec notes these as being deprecated with no specific timeline. For now, issues are overridden when not worth fixing. Will update if an exception list mechanism replaces them.

| **Owners** | EngSys (spec-gen-sdk), language teams (emitters), azsdk-cli team |
| **Reviewer ask** | Confirm two-stage pipeline and error reporting gaps |

#### Happy path

1. Spec PR merges → pipeline triggers automatically
2. **Stage A**: Create or find release plan work item
3. **Stage B**: For each language: tsp-client syncs → emitter generates → customizations applied → build → test → metadata updated → SDK PR created and linked to release plan

#### Failure paths

| Failure | Signal | Owner | Next action | Resolution |
|---------|--------|-------|-------------|------------|
| Generation failure (any language) | Failed pipeline check (buried in logs) | EngSys / language owner | Investigate logs manually | Manual |
| Customization drift | Build failure in SDK PR | azsdk-cli team | `auto-sdk-build-fix` label → auto-repair | Automatic |
| Release plan creation failure | Pipeline failure | azsdk-cli team | Investigate DevOps connectivity | Manual |

<details>
<summary>Deep spec: tools, contracts, and unresolved questions</summary>

#### Tool contract

| Tool | Role | Owner |
|------|------|-------|

> | tsp-client | Syncs TypeSpec project into SDK repo | EngSys |

Reviewer 6: JFYI I own the tool, but eng sys owns the integration with the language repos

Author: Updated: now shows @catalinaperalta (tool) and EngSys (language repo integration).

3. Architects review and approve
4. SDK PR approved and merged

#### Failure paths

| Failure | Signal | Owner | Next action | Resolution |
|---------|--------|-------|-------------|------------|
| Custom code drift | Build failure | azsdk-cli team | `auto-sdk-build-fix` label → Copilot agent auto-repairs | Automatic |
| Other CI failure | Build/test/lint red | Language owner | Pipeline troubleshooting agent diagnoses | Manual (agent-assisted) |
| API review feedback | Architect comments | Author | Resolve via TypeSpec changes → re-generate → new commit → CI re-runs | Manual |
| API review not approved | Architect rejects API surface | Author | Revise TypeSpec design, update spec PR, re-generate SDK. May require follow-up architect discussion. Not all languages auto-approve — rejection path varies per language. | Manual |
| SDK breaking change | Detection report | @raych1 / @catalinaperalta | Review and approve or fix | Manual |

<details>
<summary>Deep spec: tools, contracts, and unresolved questions</summary>

#### Tool contract

| Tool | Role | Owner |
|------|------|-------|
| Language CI pipelines | Build, test, lint, package validation | Language teams |

> | SDK breaking change detector | Detects breaking changes in generated SDK API surface. Being combined into validation check. | @raych1 / @catalinaperalta |

Reviewer 6: ```suggestion
| SDK breaking change detector | Detects breaking changes in generated SDK API surface. Being combined into validation check. | @raych1 |
```

Author: Applied Ray's updated wording. Owner is now @chunyu3 / @raych1. Removed PR link per earlier feedback.

| Language release pipelines | Publish to PyPI, Maven, npm, NuGet, Go module proxy | Language teams |
| Service Tree integration | Mark service KPIs as completed | EngSys |

#### Gap

Two processes, different maturity:
1. **SDK PR readiness** (before merge) — Multiple gaps tracked in [#15705](https://github.com/Azure/azure-sdk-tools/issues/15705): linter failures (especially samples), recorded test failures, merge conflicts, .NET-specific gaps, pipeline provisioning delay. Changelog: mgmt auto-generated reliably, data-plane not reliable.
2. **Release trigger** (after merge) — Auto-trigger on SDK PR merge being built by @raych1. Once merged, changelog/metadata are already done — just need to trigger release.

Manual approval gate on release pipeline cannot be removed for security (ARM approval = Shanghai team).

#### Open questions

- [ ] What triggers auto-release after SDK PR merge?
- [ ] For patch releases — what triggers the workflow differently?

</details>

---

## Cross-cutting contracts

### Label contract

> #### Spec PR Labels (`azure-rest-api-specs`)

Reviewer 6: not to sound like a broken record, but the PublishToCustomers label + the suppression labels should be added here

Author: Suppression labels are now noted as pending (new label TBD). For PublishToCustomers: is this still required? If yes, can you or Ray add it to the label contract table? We left it out pending confirmation.

edit: added

## Table of Contents

- [Quick Start](#quick-start)
  - [For Service Teams](#for-service-teams)
  - [For EngSys / SDK Team](#for-engsys--sdk-team)
- [Definitions](#definitions)
- [Background / Problem Statement](#background--problem-statement)
- [Goals](#goals)
- [Participants & Roles](#participants--roles)
- [Workflow Design](#workflow-design)
  - [Workflow Stages](#workflow-stages)
    - [High-Level Sequence Diagram](#high-level-sequence-diagram-service-team-journey-spec-pr-entry-point)
  - [Labels & Automation Triggers](#labels--automation-triggers)
  - [Related Process Documentation](#related-process-documentation-external-links)
  - [Architecture: Skill Chaining with Next Steps](#architecture-skill-chaining-with-next-steps)
- [Known Gaps](#known-gaps)
- [Success Criteria](#success-criteria)
- [Exceptions and Limitations](#exceptions-and-limitations)
- [Open Questions](#open-questions)
- [Appendix: Detailed Flowchart](#appendix-detailed-flowchart)

---

## Quick Start

> ### For Service Teams

Reviewer 5: Reading this, let's list the waiting points in the process where we think a human should be there:
- Namespace approval
- Spec review approval. Could be reduced to suppression review, and auto-approve with good linters
- SDK PR approval. Could be reduced to nothing once we get good at changelog, versionning, etc. For data-plane, this includes possibly an archboard or email validation, and APIView approval as well
- Release pipeline approval. Difficult to remove

Author: Added human gates callout box listing all approval points with notes on which can be reduced over time.

1. **Author TypeSpec** — Write your `.tsp` files and `tspconfig.yaml` locally (or use the `azure-typespec-author` agent skill)
2. **Open a spec PR** — Push to `azure-rest-api-specs` and open a PR. CI validates automatically.
3. **Wait for approvals** — Namespace approval (if new package), ARM review (if ARM spec), and API review must pass before merge
4. **Spec PR merges → SDK generation is automatic** — Release plan is created, SDKs are generated, and SDK PRs are opened in each language repo automatically ⚠️ *Caveat: generation failures currently fail silently — see [Known Gap #4](#known-gaps-and-next-steps)*
5. **Address SDK PR feedback** — If API review has suggestions or CI fails, resolve them (auto-repair handles custom code drift)

> 6. **Release** — Once SDK PRs merge, trigger release (or auto-release if configured). Packages publish, release plan completes.

Reviewer 5: Additional steps, get APIView approval. While we want to revisit this, this will stay for enough time that it should be mentionned.

Today, all data-plane and C# ARM requires explicit APIView approval

Author: Added APIView approval as explicit step in service team quick-start and human gates callout.

> 6. **Release** — Once SDK PRs merge, trigger release (or auto-release if configured). Packages publish, release plan completes.

Reviewer 5: Today, the release pipeline has an approval gate as well. It seems we won't be able to waive it for security reasons

Author: Noted release pipeline approval gate throughout. Can't remove for security; ARM approval = Shanghai team.

> 6. **Release** — Once SDK PRs merge, trigger release (or auto-release if configured). Packages publish, release plan completes.

Reviewer 7: you may refer this page about the responsibilities for SDK team and service team for mgmt. plane release: https://eng.ms/docs/products/azure-developer-experience/plan/mgmt-sdk-release-process

Author: Added mgmt plane release link in quick-start and Related Process Documentation table. Thanks!

> ### For EngSys / SDK Team

Reviewer 5: I'd like to separate the quick start of EngSys, and reviewers.

Author: Split into 3 sections: Service Teams, Reviewers, EngSys/SDK Team.

1. **Monitor pipelines** — Spec PR validation, SDK generation, and SDK CI pipelines run automatically
2. **Label routing** — Labels like `BreakingChangeReviewRequired`, `<lang>-namespace-pending`, and `auto-sdk-build-fix` trigger review routing and automation
3. **Generation failures** — When SDK generation fails, diagnose via pipeline logs (structured error reporting is a [known gap](#known-gaps))
4. **Auto-repair** — `auto-sdk-build-fix` label triggers Copilot cloud agent to fix custom code drift on SDK PRs
5. **Release coordination** — `azsdk_release_sdk` checks readiness; release pipelines publish to package registries

> 6. **Track progress** — [Release plan dashboard](https://aka.ms/azsdk/releaseplan-dashboard) shows where each service is in the process

Reviewer 5: Feels relevant for service team as well. It depends how quick you want the quickstart section to be :)

Author: Moved pipeline monitoring into service team quick-start (step 5).

---

## Definitions

- **TypeSpec**: Language for describing cloud service APIs. See [typespec.io](https://typespec.io).
- **SDK**: Client libraries generated from TypeSpec for .NET, Java, JavaScript, Python, Go.
- **Release Plan**: Azure DevOps work item tracking end-to-end SDK release across languages. Managed by azsdk-cli tooling.
- **API Spec PR**: Pull request in `azure-rest-api-specs` or `azure-rest-api-specs-pr` containing TypeSpec changes.

> - **SDK PR**: Pull request in a language SDK repo (e.g., `azure-sdk-for-python`) with generated code.

Reviewer 5: Not just generated code, could be customization as well. Given we're in the definition section, I think it's fair to be precise.

Author: Updated: SDK PR = generated and customized SDK code.

- **APIView**: Web tool for reviewing SDK public API surface.
- **tspconfig.yaml**: Configuration in TypeSpec project specifying emitter settings per language.
- **tsp-location.yaml**: Configuration in SDK repos pointing to the source TypeSpec project.

> - **`@azure-tools/typespec-breaking-change`**: TypeSpec-native breaking change detector. Phase A: same-version regression (any diff = error). Phase B: cross-version evolution (request narrowing / response widening = breaking).

Reviewer 5: This is one of the TypeSpec tool, but not the only one. We have linter validation as well. It depends on how you use that later in the doc, if we need to clarify the definition

Author: Updated: added TypeSpec linter as separate tool + added TypeSpec Lintdiff definition.

> - **Suppression Decorators**: `@approvedBreakingChange` (Phase B) and `@approvedUnversionedChange` (Phase A) — inline approval with auto-labeling.

Reviewer 5: I'm confused what this is, where this is applied, etc. The definition section should not need to read the doc to understand what's it's defining

Author: Rewrote spec-gen-sdk definition: what it is, where it runs (CI dry-run + merge), what it does. Also added TypeSpec Lintdiff definition.

1. **Fragmented tooling**: Different tools for TypeSpec authoring, spec validation, breaking change detection, SDK generation, and release management — developed independently with limited integration.
2. **Manual coordination**: Teams manually track progress across multiple repositories (`azure-rest-api-specs`, `azure-sdk-for-*`) and systems (Azure DevOps, GitHub, Service Tree). The [release plan dashboard](https://aka.ms/azsdk/releaseplan-dashboard) provides visibility but doesn't drive automation.
3. **Silent failures**: SDK generation failures are not surfaced to users — tools fail without actionable guidance.
4. **Scattered documentation**: Process is documented across EngHub, Wiki, and repo specs with no single authoritative source.
5. **No unified CI chain**: Spec PR validation (compile → lint → breaking change → generate → build) lacks a designed end-to-end failure-handling strategy.

### What Has Changed

- The standalone "release planner" system has been **fully removed**. Release tracking is now handled directly through Azure DevOps work items managed by azsdk-cli and the [release plan dashboard](https://aka.ms/azsdk/releaseplan-dashboard).

> - `@azure-tools/typespec-breaking-change` provides TypeSpec-native breaking change detection (replacing OpenAPI-based tools for TypeSpec specs). Crystal's breaking change detection spec defines the CI integration approach.

Reviewer 5: We're missing the TypeSpec Lintdiff tool

Author: Added TypeSpec Lintdiff to definitions, Stage 2 tooling table, and labels table (Suppression-Approved).

### Why This Matters

- **Time from API definition to SDK release** is measured in weeks for experienced teams, longer for new ones.
- **Partial agent-assisted workflow** exists today via azsdk-cli MCP tools — the agent can help with individual stages but lacks a unified end-to-end orchestration that connects all stages seamlessly.

---

## Goals

1. Provide an end-to-end guided workflow from TypeSpec authoring to SDK release
2. Enable resuming the workflow from any intermediate state ⚠️ *Caveat: resume mechanism not yet defined — the agent should detect existing state (release plan, open PRs) and pick up from the appropriate step, but the specific scenarios and detection logic are a gap to be designed*
3. Automatically track release plan status throughout the workflow
4. Surface errors with actionable guidance at every failure point
5. Integrate sub-skills seamlessly for specialized tasks
6. Support all tier-1 SDK languages with appropriate scope:

> | Plane | Required Languages |

Reviewer 5: I would prefer a third colum optionnal, so the status of Go is more clear, and we can put Rust

Author: Added Optional column: Go optional for data-plane, Rust planned.

|-------|-------------------|
| **Management plane** | .NET, Java, JavaScript, Python, Go |
| **Data plane** | .NET, Java, JavaScript, Python, Go (optional) |

---

## Participants & Roles

The workflow involves multiple participants, each entering at specific stages with defined responsibilities.

**Stage overview** (detailed in [Workflow Design](#workflow-design)):

| Stage | Name | Summary |
|-------|------|---------|
| 1 | TypeSpec Authoring | Developer authors/updates TypeSpec locally, validates, opens spec PR |
| 2 | Spec PR Validation | CI validation, ARM/API review, namespace approval on spec PR |
| 3 | Automated SDK Generation | Release plan created, SDKs generated, SDK PRs opened (automatic on spec merge) |
| 4 | SDK PR Validation & API Review | SDK CI, API review, auto-repair, SDK PR merged |
| 5 | Release Coordination | Changelog, readiness check, release pipeline, publish |

| Participant | Enters at | Responsibilities | Exit Condition |
|-------------|-----------|-----------------|----------------|

> | **Service team** | Stage 1 | Author TypeSpec, open spec PR, respond to review feedback, approve SDK PRs | SDK packages released |

Reviewer 5: You're making a stance that service teams need to approve SDK PR, this is an open discussion

Author: Removed 'approve SDK PRs' from service team responsibilities. Added open discussion note below the table.

> | **ARM review team** | Stage 2 | Review ARM resource model correctness on spec PR (ARM specs only) | `ARM-Approved` label applied |

Reviewer 5: ARMSignedOff

https://github.com/Azure/azure-rest-api-specs/issues?q=state%3Aopen%20label%3AARMSignedOff

Author: Fixed to ARMSignedOff. Also added it to the labels table.

> | **ARM review team** | Stage 2 | Review ARM resource model correctness on spec PR (ARM specs only) | `ARM-Approved` label applied |

Reviewer 5: You're missing the stewardship board. OR to put it differently, the data-plane API validation approval. This is changing as the stewardship board is not as funded as before, but something has to approve data-plane APIs.

Today this is this:
https://github.com/Azure/azure-rest-api-specs/issues?q=state%3Aopen%20label%3AAPIStewardshipBoard-SignedOff

Author: Added 'Data-plane API validation' row with note that stewardship board is in transition. Added APIStewardshipBoard-SignedOff label. Also added open question about data-plane review model.

| **API review architects** | Stage 2 + 4 | Review API design (spec PR: TypeSpec surface; SDK PR: generated API surface) | `api-approved` / `<lang>-api-approved` labels applied |

> | **Breaking change review team** | Stage 2 | Review breaking changes flagged by `BreakingChangeReviewRequired` label | Breaking change approved or rejected |

Reviewer 5: What are we talking about here? API breaking changes? SDK breaking changes? It seems at the spec level?

Author: Clarified as 'API (spec-level) breaking changes' in the participants table.

> | **SDK team (Shanghai)** | Stage 4 | Review generated SDK PRs — **only PRs with a release plan attached** (manually created PRs without release plan are not reviewed) | SDK PR approved & merged |

Reviewer 5: This is true only for ARM

Author: Scoped to ARM specs only in the participants table.

     │                 │                │                │                │                │
     │  [If FAIL]      │                │                │                │                │
     │  Fix & push     │                │                │                │                │
     │────────────────>│  (re-triggers) │                │                │                │
     │                 │───────────────>│                │                │                │
     │                 │                │                │                │                │
     │                 │  [If PASS] Request reviews      │                │                │
     │                 │───────────────────────────────->│                │                │
     │                 │                │                │                │                │
     │                 │                │                │  ARM review    │                │
     │                 │                │                │  (if ARM spec) │                │
     │                 │                │                │                │                │

>      │                 │                │                │  API review    │                │

Reviewer 5: You mean SDK API here?

Author: Changed to 'SDK API review' in the sequence diagram.

     │                 │                │                │                │                │
     │  Trigger release (or auto)       │                │                │       ┌───────>│
     │──────────────────────────────────────────────────────────────────────────┘        │
     │                 │                │                │                │                │
     │                 │                │                │                │   Publish pkgs │
     │                 │                │                │                │   Update KPI   │
     │                 │                │                │                │                │
     │  ✅ Done!       │                │                │                │                │
     │<────────────────────────────────────────────────────────────────────────────────────│
     │                 │                │                │                │                │
```

**Key insight**: After spec PR merge, the flow is largely **automated**. The service team only re-engages if:
- SDK CI fails and auto-repair can't fix it
- API review has feedback requiring TypeSpec changes

> - Release needs manual trigger (unless auto-release configured)

Reviewer 5: Note that the release pipeline has an approval step, that we don't seem to want to remove today. So it's technically auto-triggering the release pipeline, but approval still needed. 

That approval for ARM should be Shanghai team.

Author: Updated: release pipeline has approval gate (can't remove for security). ARM approval = Shanghai team. Noted throughout doc.

> > **Note: ARM vs Data Plane divergence** — ARM (management plane) and data plane specs follow the same high-level flow but diverge at review gates: ARM specs require ARM review team sign-off (`ARM-Review-Required` label) and have stricter resource model constraints. Data plane specs skip ARM review but still require API review. A separate process document will detail these divergences. See: [ARM review process](https://eng.ms/docs/products/azure-developer-experience/design/api-specs-pr/arm-review) | [Data plane review process](https://eng.ms/docs/products/azure-developer-experience/design/api-specs-pr/data-plane-review)

Reviewer 5: We raise good questions about who review data-plane PRs, if the stewardship board is out. And shoulld we consider that if the SDK looks fine, it means the TypeSpec is ok?

That would be an interesting statement:
- On ARM, if the RestAPI looks fine, we assume SDK is ok to ship
- On Data-Plane, if the SDK looks fine, we assume that the RestAPI is ok to mergee

Author: Added as open question in the ARM vs Data Plane with note for the pattern where ARM quality flows spec->SDK, data-plane quality flows SDK->spec.

| 4 | Steps 5–6: SDK CI + API review + repair | Manual (user triggers release) |
| 5 | Step 7 (sub-steps 7a–7d): Release | — |

| Stage | Name | Input | Output | Key Tools |
|-------|------|-------|--------|-----------|
| 1 | TypeSpec Authoring | Service requirements | `.tsp` files, `tspconfig.yaml` | TypeSpec compiler, authoring agent, breaking change tool, linter |
| 2 | Spec PR Validation | PR in `azure-rest-api-specs` | Validation checks, labels, APIView tokens | Validation pipeline, LintDiff, `@azure-tools/typespec-breaking-change`, APIView emitter, spec-gen-sdk dry-run |
| 3 | SDK Generation | Approved/merged spec | SDK PRs per language | tsp-client, language emitters, spec-gen-sdk, azsdk-cli (`generate`, `build`, `test`, `metadata`) |
| 4 | SDK PR Validation & API Review | SDK PRs in language repos | Approved & merged PRs | Language CI, SDK breaking change detector, API review, feedback resolution agent, pipeline troubleshooting agent, auto SDK PR repair |
| 5 | Release Coordination | Merged SDK PRs | Published packages, KPI update | Release plan tooling, release pipelines, Service Tree |

---

> ### Labels & Automation Triggers

Reviewer 5: We're missing labels, for instance there is nothing about suppressions

Author: Added Suppression-Approved, ARMSignedOff, APIStewardshipBoard-SignedOff labels to the table.

| `VersioningReviewRequired` | Spec | CI | Versioning review needed — blocks merge |
| `ARM-Review-Required` | Spec | CI | ARM spec — routes to ARM review team, blocks merge |
| `<lang>-namespace-pending` | Spec | CI | New namespace detected — awaiting architect approval |
| `<lang>-namespace-approved` | Spec | Architect | Namespace approved for this language |
| `namespace-approved-all` | Spec | Architect | Shortcut: approves all languages (mgmt plane) |
| `Approved-BreakingChange` | Spec | Review team | Breaking change approved — unblocks merge |
| `auto-sdk-build-fix` | SDK | CI / human | Triggers Copilot cloud agent to auto-repair custom code drift |
| `<lang>-api-approved` | SDK | Architect | API surface approved — unblocks SDK PR merge |
| `release-plan-linked` | SDK | Automation | SDK team reviews only PRs with this label |
| `ready-for-review` | SDK | GitHub Form | Triggers architect board review process |
| `needs-info` | SDK | Reviewer | Service team needs to provide more info |

**Gap**: Not all label → action connections are fully automated today. Some labels exist but routing to the correct team is manual or undocumented.

> #### Related Process Documentation (External Links)

Reviewer 5: Those are 404, I assume it means "we'll write that later" ?

Author: Fixed namespace approval link to point to the actual process doc. Added note that eng.ms links need corpnet access. Added mgmt plane release process link.

| Data plane review | [Data plane review process](https://eng.ms/docs/products/azure-developer-experience/design/api-specs-pr/data-plane-review) | Data plane review gates |
| API review (SDK) | [SDK API review process](https://eng.ms/docs/products/azure-developer-experience/design/api-review) | Architect board review flow |
| Release plan dashboard | [Release plan dashboard](https://aka.ms/azsdk/releaseplan-dashboard) | Track release progress |
| Spec repo contribution | [azure-rest-api-specs contribution guide](https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/api-scenario/README.md) | PR process, required checks |

> **Note**: This spec does not duplicate the content of these process documents. It defines how they connect in the end-to-end flow and where gaps exist between them.

---

#### Stage 1: TypeSpec Authoring

Developer writes/updates TypeSpec locally: `.tsp` files, `tspconfig.yaml`, local compilation, linting, agent-assisted feedback. The API version lifecycle (private preview → public preview → stable) determines the TypeSpec authoring approach — new specs start at preview, promotions to stable require breaking change review.

> | Tool | Role | Owner |

Reviewer 5: You miss the tool @jsquire is writing

Author: Added @jsquire's changelog/versioning tool to Stage 1 tooling table.

|------|------|-------|
| TypeSpec compiler | Compile `.tsp` files, catch syntax/type errors | TypeSpec team |
| TypeSpec authoring agent (`azure-typespec-author` skill) | Assist with ARM/data-plane patterns, Azure REST API guidelines | Haoling |
| `@azure-tools/typespec-breaking-change` | Phase A: same-version regression (any diff = error). Phase B: cross-version evolution (request narrowing / response widening = breaking). Inline suppression via decorators. Also runs in Stage 2 CI. | Mark Cowlishaw |

> | TypeSpec linter | Static guideline compliance (distinct from LintDiff) | TypeSpec team |

Reviewer 5: You miss the suppression process tight to TypeSpec Lintdiff, that is based on the linter.

@catalinaperalta

Author: Added TypeSpec Lintdiff with suppression process reference (@catalinaperalta) to Stage 2 tooling table. Also added Suppression-Approved label.

**✅ Supported today**: TypeSpec compiler, authoring agent, breaking change tool, linter all functional. Agent can assist with authoring patterns.

**⚠️ Gap**: Breaking change tool reports findings with DiffKind, source location, and suggested suppression decorator — but the user must apply fixes manually. No agent currently auto-resolves these findings.

> **🎯 Next step**: Build author-validation loop where agent auto-applies suppression decorators based on structured breaking change output.

Reviewer 5: I need more details on that to have an opinion

Author: Expanded this into the Stage 2 gaps section with more concrete detail about the validation chain gap.

---

#### Stage 2: Spec PR Validation

PR in `azure-rest-api-specs` triggers: compilation → LintDiff → breaking change detection (Phase A + B) → APIView token generation → SDK generation dry-run per language → labels applied.

| Tool | Role | Owner |
|------|------|-------|
| Spec PR validation pipeline | Orchestrates full validation suite on PR open/update | EngSys |
| TypeSpec compiler | CI compilation | TypeSpec team |

> | LintDiff | Spectral-based linting on swagger diffs | EngSys |

Reviewer 5: This is getting replaced by TypeSpec Lintdiff

Author: Updated: replaced LintDiff entry with TypeSpec Lintdiff (TypeSpec-native, replacing Swagger-based Spectral LintDiff).

| `@azure-tools/typespec-breaking-change` | Phase A + B detection. Auto-adds `BreakingChangeReviewRequired` / `VersioningReviewRequired` labels | Mark |
| APIView emitter (`typespec-apiview`) | Generates API surface token files for architect review | APIView team |
| spec-gen-sdk | SDK generation validation — ensures spec can produce SDK code | EngSys (Renhe) |

> **✅ Supported today**: Full validation pipeline runs on PR open/update. Labels auto-applied. APIView tokens generated. spec-gen-sdk dry-run validates generation.

Reviewer 5: There are more tools today, like Avocado or OAV, but they should disapear. Maybe mentions them as "they need to disapear".

This is tight to the work @timotheeguerin is doing on example and readme

Author: Added Avocado/OAV as deprecated tools being phased out in Stage 2 tooling table. Noted tie to @timotheeguerin's work on example and readme validation. Added as gap #4 in Stage 2.

**⚠️ Gaps**:
1. Validation steps run independently. No designed chain for: failure ordering, how breaking change results gate SDK generation, per-language generation failure reporting back to spec PR. No unified PR comment summarizing all results with next steps.

> 2. **No service readiness check** — no automated verification that service onboarding prerequisites are met (service registered, team configured, emitters ready) before spec PR can merge. This is currently a manual tribal-knowledge step.

Reviewer 5: I don't get this one

Author: Removed Gap #2 I think it was redundant with #3 (API end point readiness check/ service readiness check.

Author: Correction: removed the service readiness gap (Gap #2) as it was vague and mostly covered by existing CI validation. The real gap here is **endpoint liveness verification** (now Gap #11 in the Known Gaps table) — verifying that API endpoints are actually deployed and working before we proceed to SDK generation. @prkannap suggested this during our review discussion. Marked as aspirational since it's a hard problem.

3. **No endpoint liveness verification** — for public specs, there is no check that the API endpoint is tested and live before the spec PR is merged. If the endpoint is not deployed, the SDK will be generated for an API that doesn't work yet. This should be a pre-merge gate or at minimum a documented prerequisite.

**🎯 Next steps**:
- Design unified PR comment that aggregates all validation results with clear next actions per failure type.
- Investigate adding a service readiness pre-merge check (future feature).

> - Investigate endpoint liveness/testing verification as a pre-merge gate for public specs.

Reviewer 5: Note, this is a really hard problem, I wouldn't work on that just yet. But ok to mention it as inspirational

Author: Marked as aspirational long-term goal in Stage 1 next steps. Not a near-term priority.

**🎯 Next step**: Complete SDK breaking change integration into CI. Extend auto-repair pattern to cover more failure categories.

---

#### Stage 5: Release Coordination

Release plan work item created/updated → namespace approval (if new package) → readiness checked per language → release pipeline triggered (manual approval gate) → packages published → release plan auto-completes → Service Tree KPI updated.

| Tool | Role | Owner |
|------|------|-------|
| Release plan tooling (`azsdk_create_release_plan`, `azsdk_get_release_plan`, etc.) | Create/update/link Azure DevOps work items | azsdk-cli team |
| Release pipeline (`azsdk_release_sdk`) | Check readiness, trigger release | azsdk-cli team |
| Language release pipelines | Publish to PyPI, Maven, npm, NuGet, Go module proxy | Language teams |
| Service Tree integration | Mark service KPIs as completed | EngSys |

**✅ Supported today**: Release plan tooling (create/update/link), `azsdk_release_sdk` readiness check, language release pipelines, Service Tree KPI update, auto-release for configured packages.

> **⚠️ Gap**: Steps 7a–7c are manual today. No agent auto-triggers changelog date update or release pipeline after SDK PR merge. User must manually initiate release. Transition from Stage 4 → 5 is not agent-driven.

Reviewer 5: 7a and 7c ?

Author: Fixed — Stage 5 gap now describes two phases (readiness + release) instead of referencing individual step numbers.

> **🎯 Next step**: Investigate auto-release flow (SDK PR merge → changelog date → release trigger) for packages configured for it.

Reviewer 5: We discussed with @raych1 , that there is two steps here:
- SDK PR readiness. Create ChangeLog, fix breaking changes, etc.
- SDK release. Auto-trigger a release pipeline when the SDK PR is merged

We'll start investigating the second one now, as this is a low hanging fruit that speeds up the process already

Author: Split Stage 5 into two phases: (1) SDK PR readiness (changelog, breaking changes) and (2) SDK release (auto-trigger on merge). Noted starting with phase 2 as low-hanging fruit per discussion with @raych1.

---

## Success Criteria

This workflow is complete when:

- [ ] Users can complete full TypeSpec → SDK release workflow with agent guidance
- [ ] Agent detects existing state and resumes from appropriate step
- [ ] Release plan is automatically updated at each step
- [ ] All sub-skills integrate seamlessly without context loss between stages
- [ ] Local and pipeline SDK generation paths both work
- [ ] Breaking change findings from CI are surfaced clearly to the user
- [ ] APIView feedback can be resolved within the workflow
- [ ] Works for all tier-1 SDK languages (per language scope table)
- [ ] Errors at every stage produce structured, actionable guidance
- [ ] Service Tree KPI is updated on release completion

---

## Exceptions and Limitations

### Exception 1: Architect Board Review

> **Description**: First GA releases require architect board review (human decision-making outside automated workflow). Preview releases do not require review but can request one.

Reviewer 5: Technically first preview requires namespace approval. This could be mentionned here.

Author: Added: first preview requires namespace approval in the release type table (separate from GA first release).

- [Goals](#goals)
- [Participants & Roles](#participants--roles)
- [Workflow Design](#workflow-design)
  - [Workflow Stages](#workflow-stages)
    - [High-Level Sequence Diagram](#high-level-sequence-diagram-service-team-journey-spec-pr-entry-point)
  - [Labels & Automation Triggers](#labels--automation-triggers)
  - [Related Process Documentation](#related-process-documentation-external-links)
  - [Architecture: Skill Chaining with Next Steps](#architecture-skill-chaining-with-next-steps)
- [Known Gaps](#known-gaps)
- [Success Criteria](#success-criteria)
- [Exceptions and Limitations](#exceptions-and-limitations)
- [Open Questions](#open-questions)
- [Appendix: Detailed Flowchart](#appendix-detailed-flowchart)

---

## Quick Start

### For Service Teams

1. **Author TypeSpec** — Write your `.tsp` files and `tspconfig.yaml` locally (or use the `azure-typespec-author` agent skill)
2. **Open a spec PR** — Push to `azure-rest-api-specs` and open a PR. CI validates automatically.
3. 🧑‍💻 **Wait for approvals** — The following require human sign-off before spec PR can merge:
   - **Namespace approval** (first preview only) — Architect approves new package namespaces

>    - **Spec review** — ARM review (ARM specs) + API review via APIView (all data-plane + C# ARM). Could eventually be reduced to suppression review + auto-approve with better linters.

Reviewer: APIView will be replaced with APIReviewHub generated review PRs on Github

4. **Spec PR merges → SDK generation is automatic** — Release plan is created, SDKs are generated, and SDK PRs are opened in each language repo automatically ⚠️ *Caveat: generation failures currently fail silently — see [Known Gap #4](#known-gaps)*

> 5. 🧑‍💻 **SDK PR review & approval** — SDK CI runs automatically. If API review has suggestions or CI fails, resolve them (auto-repair handles custom code drift). For management plane, Shanghai team reviews SDK PRs with release plans. APIView approval required for generated SDK API surface — review requests use a [GitHub Form template](https://github.com/Azure/azure-sdk/blob/main/.github/workflows/src/arch-board-review/ARCH-BOARD-REVIEW-PROCESS.md) (replaces email). Could eventually be reduced once changelog/versioning automation matures.

Reviewer: API approved is required. This will be via APIView (legacy) or APIReviewHub (GH-based)

6. 🧑‍💻 **Release** — Once SDK PRs merge, trigger release pipeline (**manual approval gate** required for security; difficult to remove. ARM approval = Shanghai team). Packages publish, release plan completes. See also: [Management plane release responsibilities](https://eng.ms/docs/products/azure-developer-experience/plan/mgmt-sdk-release-process)

### For Reviewers

1. **ARM review** (ARM specs only) — Review resource model correctness on spec PRs; apply `ARMSignedOff` label

> 2. **API review** — Review API design via APIView tokens on spec PRs and SDK PRs; apply `api-approved` / `<lang>-api-approved` labels

Reviewer: There is no need for these labels, unless they are informational only. API approval is tracked on APIView/APIReviewHub. The labels cannot be authoritative.

Author: So what actually gates merge?

| **Breaking change review team** | Stage 2 | Review API (spec-level) breaking changes flagged by `BreakingChangeReviewRequired` label (ARM specs only) | Breaking change approved or rejected |
| **SDK team (Shanghai)** | Stage 4 | Review generated SDK PRs — **only PRs with a release plan attached** (manually created PRs without release plan are not reviewed) | SDK PR approved & merged |
| **EngSys team** | Stages 2–5 | Pipeline automation, tooling, generation infrastructure | Pipelines green |
| **Language owners** | Stage 3–4 | Maintain per-language generation scripts, CI, emitter config | Language CI passes |
| **Namespace approvers** | Stage 2 | Approve new package namespaces (extracted from `tspconfig.yaml`) | `<lang>-namespace-approved` labels applied |

> **Open discussion**: Whether service teams should be required to approve SDK PRs is still under discussion. The current flow does not require service team approval on SDK PRs, but this may change.

### Entry conditions per participant

- **Service team**: Has service requirements and access to `azure-rest-api-specs` repo
- **ARM review team**: Spec PR has ARM changes + CI passes + `ARM-Review-Required` label present

> - **API review architects**: APIView tokens generated (spec stage) or SDK PR open (SDK stage)

Reviewer: With ARH, the review PR must be opened. However that should slot into this. Unlike APIView, ARH PRs will not be created automatically as they are in APIView.

│  Team    │     │  (GH)    │     │Pipeline  │     │(ARM/API) │     │(per lang)│     │ Pipeline │
└────┬─────┘     └────┬─────┘     └────┬─────┘     └────┬─────┘     └────┬─────┘     └────┬─────┘
     │                 │                │                │                │                │
     │  Open spec PR   │                │                │                │                │
     │────────────────>│                │                │                │                │
     │                 │                │                │                │                │
     │                 │  PR triggers   │                │                │                │
     │                 │───────────────>│                │                │                │
     │                 │                │                │                │                │
     │                 │                │── Compile      │                │                │
     │                 │                │── LintDiff     │                │                │
     │                 │                │── Breaking chg │                │                │

>      │                 │                │── APIView gen  │                │                │

Reviewer: Somewhere in here we need to include APIReviewHub. It could be that opening the SDK PR triggers opening of the API review PR

Author: Is it right to say currently it is NOT auto-created (unlike APIView). So should we specify how is it triggred, Is it SDK PR creation? Manual action?

**🎯 Next step**: Build author-validation loop where agent auto-applies suppression decorators based on structured breaking change output. Long-term aspirational goal: agent detects existing workflow state and resumes from the appropriate step (hard problem — ok to mention as inspirational, not a near-term priority).

---

#### Stage 2: Spec PR Validation

PR in `azure-rest-api-specs` triggers: compilation → LintDiff → breaking change detection (Phase A + B) → APIView token generation → SDK generation dry-run per language → labels applied.

| Tool | Role | Owner |
|------|------|-------|
| Spec PR validation pipeline | Orchestrates full validation suite on PR open/update | EngSys |
| TypeSpec compiler | CI compilation | TypeSpec team |
| TypeSpec Lintdiff | TypeSpec-native linting on PR diffs — validates guideline compliance, replacing older Swagger-based Spectral LintDiff. Includes a suppression process for known exceptions (linter-based, see @catalinaperalta). | EngSys / TypeSpec team |
| `@azure-tools/typespec-breaking-change` | Phase A + B detection. Auto-adds `BreakingChangeReviewRequired` / `VersioningReviewRequired` labels | Mark |

> | APIView emitter (`typespec-apiview`) | Generates API surface token files for architect review | APIView team |

Reviewer: With ARH, this will likely simply be retired. TypeSpec PRs are ALREADY on GitHub and since they are already API surface areas, there's no "implementation" to strip out.

CI runs (build → test → lint → package validation → SDK breaking change detection). If build fails on custom code drift: `auto-sdk-build-fix` label → Copilot cloud agent auto-repairs → commits fix → CI re-runs. API review surfaces SDK public API diff for architect review → architects leave comments → developer resolves via TypeSpec changes (re-triggers generation) → iterate until approved and merged.

| Tool | Role | Owner |
|------|------|-------|
| Language CI pipelines | Build, test, lint, package validation | Language teams |
| SDK breaking change detector | Detects breaking changes in generated SDK API surface. Being combined into SDK validation check. Could the auto-repair pattern (below) also apply here for resolution? | Ray & Crystal |
| APIView (current) | SDK public API surface review — architects review and approve via web UI | APIView team |
| **API Review Hub** (replacing APIView) | GitHub-native API review workflow: creates automated review PRs showing API diffs, architects review in GitHub, approval state tracked in Cosmos DB, release-gating via `GET /api/releases/check-gate`. Uses `azure-sdk-automation` GitHub App. | @tjprescott |

> | APIView feedback resolution agent (`azsdk-common-apiview-feedback-resolution` skill) | Helps resolve API review comments via TypeSpec changes | azsdk-cli team |

Reviewer: As a gap: we will need to evaluate whether this can be updated to work with ARH's GH-based reviews. I would assume it would since Github should be easier to work with than APIView, but it's still untested.

| Pipeline troubleshooting agent (`azsdk-common-pipeline-troubleshooting` skill) | Diagnoses CI failures | azsdk-cli team |
| Auto SDK PR repair (`azsdk_customized_code_update` + Copilot cloud agent) | When custom code drifts and breaks the build, `auto-sdk-build-fix` label triggers a Copilot cloud agent to fix custom code, regenerate, and rebuild. Shared orchestration in `eng/common/`, per-language opt-in. | azsdk-cli team |

**✅ Supported today**: Language CI pipelines, APIView feedback resolution agent, pipeline troubleshooting agent, auto SDK PR repair (label-triggered). SDK breaking change detection being integrated. API Review Hub has end-to-end prototype working for Python.

> **⚠️ Gap**: SDK breaking change detection integration in progress (being combined into validation check). Auto-repair only handles custom-code drift — not all CI failure types. API review routing relies on manual label application in some cases. API Review Hub is not yet in production — transition from APIView to API Review Hub is in progress.

Reviewer: Why does API review rely in any way on labels? Is this true for data plane? Mgmt plane? Both?

| 11 | No endpoint liveness verification before spec PR merge | 2 | For public specs, no check that the API endpoint is tested and live before merge. SDK may be generated for an undeployed API. | Pre-merge gate or documented prerequisite verifying endpoint is tested and live | TBD |
| 12 | No auto-release after SDK PR merge | 5 | Auto SDK PR creation is actively being built (release plan → SDK generation), but auto-release after merge is undefined. This is the last missing piece for full end-to-end automation (spec merge → release plan → SDK generation → SDK PR → merge → publish). | SDK PR merge triggers changelog date update + release pipeline automatically for configured packages | TBD |

---

## Success Criteria

This workflow is complete when:

- [ ] Users can complete full TypeSpec → SDK release workflow with agent guidance
- [ ] Agent detects existing state and resumes from appropriate step
- [ ] Release plan is automatically updated at each step
- [ ] All sub-skills integrate seamlessly without context loss between stages
- [ ] Local and pipeline SDK generation paths both work
- [ ] Breaking change findings from CI are surfaced clearly to the user

> - [ ] APIView feedback can be resolved within the workflow

Reviewer: ```suggestion
- [ ] API review feedback can be resolved within the workflow
```

This way you cover APIView and ARH

              │ • SDK generation dry-run (spec-gen-sdk)           │
              │ • Labels applied                                  │
              └───────────────────────┬───────────────────────────┘
                                      │
           ┌──────────────────────────┴──────────────────────────┐
           │ FAIL                                                │ PASS
           ▼                                                     ▼
┌─────────────────────────┐                     ┌─────────────────────────┐
│ Fix issues, push to PR  │                     │ All checks pass         │
│ (re-triggers pipeline)  │                     └───────────┬─────────────┘
└─────────────────────────┘                                 │
                                                           ▼
             ┌───────────────────────────────────────────────────┐
             │  STEP 4b: ARM Review & API Review (offline)      │
             │  • ARM review (ARM specs only — architect signs   │
             │    off on resource model correctness)             │
             │  • API review via APIView (all specs — reviewer   │

>              │    approves API design based on generated tokens) │

Reviewer: With ARH we are unlikely to support TypeSpec since the actual spec PR already is a GH PR. We don't "release" TypeSpecs, so release gating is irrelevant. Similiar to the namespace approvals, the review should occur on the TypeSpec PR and approval is reflected in merging the PR.

│ linked to release plan  │                     │ retry or escalate       │
│ automatically           │                     │ (loop back to fix       │
└───────────┬─────────────┘                     │  TypeSpec & retry)      │
            │                                   └─────────────────────────┘
                                      │
                                      ▼
╔═══════════════════════════════════════════════════════════════════╗
║  STAGE 4: SDK PR Validation & API Review                         ║
╚═══════════════════════════════════════════════════════════════════╝
                                    │
              ┌───────────────────────────────────────────────────┐
              │  STEP 5: SDK PR CI Pipeline                      │
              │  (auto-triggers on SDK PR open/update)           │
              └───────────────────────┬───────────────────────────┘
                                      │
                                      ▼
              ┌───────────────────────────────────────────────────┐
              │ • Build → Test → Lint → Package validation       │
              │ • SDK breaking change detection                   │

>               │ • APIView generated for SDK public API surface    │

Reviewer: ```suggestion
              │ • API review generated for SDK public API surface    │
```

Covers both APIView and ARH

                                      │
                                      ▼
              ┌───────────────────────────────────────────────────┐
              │  7a. Update changelog with release date (MANUAL) │
              │  • Replace "Unreleased" → version + date         │
              │  • Agent can help via azsdk_package_update_       │
              │    changelog_content, but user must trigger       │
              └───────────────────────┬───────────────────────────┘
                                      │
                                      ▼
              ┌───────────────────────────────────────────────────┐
              │  6b. Check release readiness (MANUAL)            │
              │  • azsdk_release_sdk checks readiness per lang   │
              │  • Namespace approval (if new package)           │
              └───────────────────────┬───────────────────────────┘
                                      │
                                      ▼
              ┌───────────────────────────────────────────────────┐
              │  7c. Trigger release pipeline                    │
              │  • Manual approval gate (user triggers)          │

>               │  • OR auto-release if package is configured      │

Reviewer: Somewhere in here we should reflect the automated release gates. These currently come from APIView but soon will accept either APIView or ARH and then eventually will accept only ARH

## Table of Contents

- [Quick Start](#quick-start)
  - [For Service Teams](#for-service-teams)
  - [For Reviewers](#for-reviewers)
  - [For EngSys / SDK Team](#for-engsys--sdk-team)
- [Definitions](#definitions)
- [Background / Problem Statement](#background--problem-statement)
- [Goals](#goals)
- [Participants & Roles](#participants--roles)
- [Workflow Design](#workflow-design)
  - [Workflow Stages](#workflow-stages)
    - [High-Level Sequence Diagram](#high-level-sequence-diagram-service-team-journey-spec-pr-entry-point)
  - [Labels & Automation Triggers](#labels--automation-triggers)
  - [Related Process Documentation](#related-process-documentation-external-links)
  - [Architecture: Skill Chaining with Next Steps](#architecture-skill-chaining-with-next-steps)
- [Known Gaps](#known-gaps)
- [Success Criteria](#success-criteria)
- [Exceptions and Limitations](#exceptions-and-limitations)
- [Open Questions](#open-questions)

> - [Appendix: Detailed Flowchart](#appendix-detailed-flowchart)

Reviewer 8: I think we also need a section in here that ties the "no spec change => release" workflow into this.   They should have a similar flow and at least have the same requirements/release gates for partners.

The data shows that this is right around the 50% case for both Python and .NET.   I'd imagine it is similar for the other languages.

2. **SDK release** (Step 7c) — Auto-trigger release pipeline when SDK PR is merged (low-hanging fruit being investigated now with @raych1)

No agent auto-triggers changelog date update or release pipeline after SDK PR merge. Release pipeline has a manual approval gate (cannot be removed for security reasons; for ARM, Shanghai team approves).

**🎯 Next step**: Start with phase 2 (auto-trigger release pipeline on SDK PR merge) as it's simpler and speeds up the process immediately. Phase 1 (SDK PR readiness automation) follows.

**Release type approval differences**:

| Release Type | Approval Gates | Notes |
|-------------|----------------|-------|
| Preview (first) | Namespace approval required | Fastest path; namespace approval needed for new packages |
| Preview (update) | No architect board review (can be requested) | Fastest path |
| GA (first release) | Architect board review required | Namespace approval needed for new packages |

> | GA (update) | Standard review | Breaking changes require separate approval |

Reviewer 8: What is "standard review" in this context?

| 8 | Scattered documentation | All | Process docs in EngHub, Wiki, repo specs | Single authoritative source | Sameeksha + Praveen |
| 9 | Release is a two-phase manual process | 5 | Phase 1 (SDK PR readiness: changelog, breaking changes) and Phase 2 (release trigger) both require manual action | Phase 2 first: auto-trigger release pipeline on SDK PR merge. Phase 1 later: automate changelog + readiness | @raych1 / TBD |
| 10 | ARM vs data plane process divergence undocumented | 2, 4 | ARM and data plane follow same high-level flow but diverge at review gates; no single doc captures differences | Linked process docs for each plane with clear divergence points | Sameeksha + Praveen |
| 11 | No endpoint liveness verification before spec PR merge | 2 | For public specs, no check that the API endpoint is tested and live before merge. SDK may be generated for an undeployed API. | Pre-merge gate or documented prerequisite verifying endpoint is tested and live | TBD |

> | 12 | No auto-release after SDK PR merge | 5 | Auto SDK PR creation is actively being built (release plan → SDK generation), but auto-release after merge is undefined. This is the last missing piece for full end-to-end automation (spec merge → release plan → SDK generation → SDK PR → merge → publish). | SDK PR merge triggers changelog date update + release pipeline automatically for configured packages | TBD |

Reviewer 8: This also needs definition for what happens if the API is not approved.  We're making the assumption here that it is automatic, and that will not always be the case across all languages.

> **See also**: [SDK PR Release Readiness tracking issue (#15705)](https://github.com/Azure/azure-sdk-tools/issues/15705) for the full consolidated gap list including .NET-specific gaps, ESRP publishing failures, and network isolation policy impacts.

---

## Success Criteria

This workflow is complete when:

- [ ] Users can complete full TypeSpec → SDK release workflow with agent guidance
- [ ] Agent detects existing state and resumes from appropriate step
- [ ] Release plan is automatically updated at each step
- [ ] All sub-skills integrate seamlessly without context loss between stages
- [ ] Local and pipeline SDK generation paths both work
- [ ] Breaking change findings from CI are surfaced clearly to the user
- [ ] APIView feedback can be resolved within the workflow
- [ ] Works for all tier-1 SDK languages (per language scope table)
- [ ] Errors at every stage produce structured, actionable guidance
- [ ] Service Tree KPI is updated on release completion

---

## Exceptions and Limitations

### Exception 1: Architect Board Review

> **Description**: First GA releases require architect board review (human decision-making outside automated workflow). First preview releases require namespace approval for new packages. Subsequent preview releases do not require review but can request one. See updated [API review process](https://github.com/Azure/azure-sdk/blob/main/.github/workflows/src/arch-board-review/ARCH-BOARD-REVIEW-PROCESS.md).

Reviewer 8: This is true for more than just first GA release.   All GA releases should expect architect review, even if some languages choose to opt out for some scenarios - the workflow must account for it.

**Impact**: Workflow cannot fully automate GA approval or first preview namespace approval.

**Status**: Three workstreams converging on this:
1. **GitHub Forms + Actions (PR #10037, shipped)** — Service teams submit review requests via GitHub Form in `azure-sdk` repo (replacing email). `arch-board-review.yml` template is an explicit **bridge** until API Review Hub ships. `namespace-review.yml` stays long-term. `approval-close.yml` validates authorized approvers and auto-closes issues.
2. **API Review Hub (PR #15789, in progress)** — Replaces APIView with synthetic GitHub PRs in language repos. When this ships, `arch-board-review.yml` can be retired. Namespace approvals remain out of scope for API Review Hub.

> 3. **Spec PR-based namespace approval (longer-term vision)** — Namespace approval moves to the spec PR itself. When spec PR merges, namespace is considered approved. Would eventually retire `namespace-review.yml` too and simplify E2E: one approval gate (spec merge) unlocks everything.

Reviewer 8: We're over-indexing on a spec creation/change triggering.   Data shows that this is only the 50% case.   We have just as many releases that take place without a spec change.

While the majority of our packages are for REST services, there are a non-trivial amount that have no spec whatsoever - such as the messaging services, companion packages for language ecosystems and such.   It's true these are minorities, but the process must account for them as they also need package+namespace approval.

**Impact**: .NET is ahead on auto-repair and linter integration. Other languages can adopt the same patterns.

**Next step**: Schedule alignment meeting to document remaining .NET-specific tools and confirm integration points.

---

## Open Questions

- [ ] **Q1**: Does `@azure-tools/typespec-breaking-change` CI output provide enough structured context (DiffKind, source location, suggested suppression decorator text) for an agent to auto-resolve? Need confirmation from Mark.
  - Options: (a) Output is already sufficient, (b) Need additional structured output format

- [ ] **Q2**: How should `BreakingChangeReviewRequired` / `VersioningReviewRequired` labels route to the correct review team? Is this existing GitHub CODEOWNERS-based routing or new automation?
  - Options: (a) CODEOWNERS, (b) Custom GitHub Action, (c) Azure DevOps integration

- [ ] **Q3**: What is the .NET team's tooling stack and where are the integration points with azsdk-cli?

>   - Context: .NET team has developed independent tooling. Need alignment on shared infrastructure.

Reviewer 8: The important call-out here is "complimentary" tooling, and I'd very much appreciate if we are mindful of how we represent this.

We've been very focused on ensuring alignment with the inner loop work and azsk-cli tooling.   Nothing should be competing or conflicting.  Those areas with cross-language potential have been surfaced for discussion and integration.

    - Auto SDK PR repair — shared orchestration in `eng/common/`, per-language opt-in via thin trigger workflow + `SDK_BUILD_REPAIR_ENABLED` flag
    - Safety validation gate — shared cross-language denylist + per-language build check
    - Copilot cloud agent — triggered by `auto-sdk-build-fix` label on failing SDK PRs
  - **Remaining question**: Are there .NET-specific tools beyond `azsdk_customized_code_update` that should integrate into the shared workflow?
  - **Next step**: Schedule alignment meeting with .NET team.

- [ ] **Q4**: Should spec-gen-sdk generation failures be reported as PR comments, or as structured data that the agent can parse and act on?
  - Options: (a) PR comments only, (b) Structured JSON output + PR comment summary, (c) Both

- [ ] **Q5**: For Patch releases — what triggers the workflow? Is it different from Preview/GA?

---

## Appendix: Detailed Flowchart

The following detailed flowchart shows every step, decision branch, and gap in the end-to-end flow. For a high-level view, see the [sequence diagram](#high-level-sequence-diagram-service-team-journey-spec-pr-entry-point) above.

```

>                          ┌─────────────────────────────────┐

Reviewer 8: I'd recommend telling copilot to move to a mermaid diagram here.   It'll be easier to get the formatting consistent.

- `@azure-tools/typespec-breaking-change` provides TypeSpec-native breaking change detection (replacing OpenAPI-based tools for TypeSpec specs). Crystal's breaking change detection spec defines the CI integration approach.

### Why This Matters

- **Time from API definition to SDK release** is measured in weeks for experienced teams, longer for new ones.
- **No agent-assisted end-to-end workflow** exists today — each stage requires manual tool invocation and knowledge of the next step.

---

## Goals

1. Provide an end-to-end guided workflow from TypeSpec authoring to SDK release
2. Enable resuming the workflow from any intermediate state
3. Automatically track release plan status throughout the workflow
4. Surface errors with actionable guidance at every failure point
5. Integrate sub-skills seamlessly for specialized tasks
6. Support all tier-1 SDK languages with appropriate scope:

| Plane | Required Languages |
|-------|-------------------|
| **Management plane** | .NET, Java, JavaScript, Python, Go |
| **Data plane** | .NET, Java, JavaScript, Python |

---

## Workflow Design

### Overview

> The workflow is an orchestration layer that:

Reviewer 9: hello @samvaity I have some questions to confirm:
1. who will run the workflow? is it copilot agent? 
2. And how to store the status of the workflow such as current stage working on? will we store all those status in Devops release work item?
3. And and a release workflow may last a long time (maybe several days) how to resume the workflow? by agent itself or human manual prompt?

Author: **1. Who runs the workflow?**
The user drives it via GitHub Copilot agent, which invokes azsdk-cli MCP tools. The user is in the loop at every step.

**2. How is workflow status stored?**
The Azure DevOps release plan work item is the status store. It tracks overall status (`Status`, `AttestationStatus`, `ActiveSpecPullRequest`) and per-language status (`SDKInfo` with `GenerationStatus`, `ReleaseStatus`, `PullRequestStatus`, `SdkPullRequestUrl`). 

**3. How to resume?**
The previous design proposed "Resumable Workflow Scenarios" and some of that resume logic already exists e.g., `azsdk_get_release_plan` returns the full state (linked PRs, per-language status), and when spec is approved it returns NextSteps: "Run SDK generation". 
The user starts a new conversation and provides a release plan ID or TypeSpec path. The agent queries the release plan, sees the current state, and picks up from there. It's manual prompt to resume, but the agent uses the release plan state to figure out where to continue.

1. **Identifies user intent** and gathers service details
2. **Assesses current state** (existing release plans, PRs, completed steps)
3. **Orchestrates sub-skills** in correct sequence
4. **Tracks progress** via release plan work items
5. **Handles failures** with troubleshooting guidance

### Workflow Stages

The end-to-end flow consists of 5 stages:

```
  Stage 1              Stage 2              Stage 3
  TypeSpec             Spec PR              SDK Generation
  Authoring            Validation           (per language)

>   (local)              (CI)

Reviewer 9: Will add 'SDK breaking change detection and mitigation' in Stage 2 'Spec PR validation' also? we want to shift 'SDK breaking change detection and mitigation' left to avoid back-and-force between spec repo and sdk repo?

Author: Yes, the stage 2 already includes spec-gen-sdk dry-run the breaking change detector tool would run as part of that flow. The updated detailed diagram (just pushed) shows Step 4 (CI Validation Pipeline) where this fits. I'll add `azsdk_package_detect_breaking_change` to the Stage 2 tools table to make this explicit.

This means SDK breaking change detection runs at **two layers** (shift-left):
- **Stage 2 (spec PR)**: Early detection + resolution via TypeSpec customization before merge -- avoids back-and-forth
- **Stage 4 (SDK PR)**: Catches anything that slipped through (e.g., suppressed at Stage 2, or emitter-specific issues only visible in built SDK)

  ┌──────┐  opens PR  ┌──────┐  merged    ┌──────┐
  │ .tsp │───────────▶│  CI  │───────────▶│ Gen  │
  │files │  (manual)  │checks│ (auto:     │build │
  └──────┘            └──────┘ spec-gen-  │test  │
                                sdk)      └──┬───┘
                                             │ SDK PRs created
                                             ▼

>                       Stage 4              Stage 5

Reviewer 9: Why Stage 3 directly to stage 5?  After 'stage 3: SDK Generation', shall we enforce Stage 4: SDK PR Validation    Release & API Review before release SDK?

Author: I've pushed an updated detailed stepwise diagram that clearly shows the flow:

**Stage 3 (SDK Generation)** → **Stage 4 (SDK PR Validation & API Review)** → **Stage 5 (Release)**

No release without CI validation and API review.

                      SDK PR Validation    Release
                      & API Review         Coordination
                      ┌──────┐  merged    ┌───────┐
                      │CI+API│───────────▶│Publish│

>                       │review│  (manual)  │  KPI  │

Reviewer 10: Maybe I miss something, may I know what does 'Publish KPI' mean?

Author: It refers to updating Service Tree KPIs. When SDK packages are published to registries, the release plan auto-completes and marks the service's SDK coverage metrics as completed in Service Tree. This is how the organization tracks which services have shipped SDKs.

From [`typespec-to-sdk.instructions.md`](https://github.com/Azure/azure-sdk-tools/blob/main/eng/common/instructions/azsdk-tools/typespec-to-sdk.instructions.md#L92):

> Release plan completion is required for KPI attestation in service tree.

The `azsdk_get_kpi_attestation_status` tool checks this status for a given product and release type.

| 4 | SDK PR Validation & API Review | SDK PRs in language repos | Approved & merged PRs | Language CI, SDK breaking change detector, APIView, feedback resolution agent, pipeline troubleshooting agent, auto SDK PR repair |
| 5 | Release Coordination | Merged SDK PRs | Published packages, KPI update | Release plan tooling, release pipelines, Service Tree |

---

#### Stage 1: TypeSpec Authoring

Developer writes/updates TypeSpec locally: `.tsp` files, `tspconfig.yaml`, local compilation, linting, agent-assisted feedback.

| Tool | Role | Owner |
|------|------|-------|
| TypeSpec compiler | Compile `.tsp` files, catch syntax/type errors | TypeSpec team |
| TypeSpec authoring agent (`azure-typespec-author` skill) | Assist with ARM/data-plane patterns, Azure REST API guidelines | Haoling |
| `@azure-tools/typespec-breaking-change` | Phase A: same-version regression (any diff = error). Phase B: cross-version evolution (request narrowing / response widening = breaking). Inline suppression via decorators. Also runs in Stage 2 CI. | Mark Cowlishaw |
| TypeSpec linter | Static guideline compliance (distinct from LintDiff) | TypeSpec team |

> **Gap**: Breaking change tool reports findings with DiffKind, source location, and suggested suppression decorator — but the user must apply fixes manually. No agent currently auto-resolves these findings.

Reviewer 10: I like TypeSpec Authoring part, it is clear! And I agree the current gap, it would be useful to create an 'author - validation/assessment' loop. Share an related issue that tracks this: https://github.com/Azure/azure-sdk-tools/issues/14529

Author: Thanks! Linked issue noted — the author-validation loop is exactly the gap we want to close.

---

#### Stage 3: SDK Generation

**Trigger**: When a spec PR is merged in `azure-rest-api-specs`, the spec-gen-sdk pipeline automatically kicks off SDK generation for each configured language and creates SDK PRs in the corresponding language repos (e.g., `azure-sdk-for-python`). Alternatively, developers can generate locally using azsdk-cli for faster iteration before opening a PR.

For each target language: tsp-client syncs TypeSpec → emitter generates code → customizations applied → build → test (playback) → metadata updated → SDK PR created. Two paths: local (developer iterates, creates PR manually) or pipeline (spec-gen-sdk generates and creates PRs automatically on spec PR merge).

| Tool | Role | Owner |
|------|------|-------|
| tsp-client | Syncs TypeSpec project into SDK repo, manages `tsp-location.yaml` | EngSys |
| Language emitters | Generate client library code from TypeSpec (one per language) | Language teams |

> | spec-gen-sdk | Pipeline automation — runs full generation workflow, creates SDK PRs | EngSys (Renhe) |

Reviewer 10: Just confirm my understanding on what is `spec-gen-sdk` tool:
Does `spec-gen-sdk` underline call the same script that each language owners wrote for below individual tool, but it is for pipeline?

<img width="387" height="128" alt="Image" src="https://github.com/user-attachments/assets/a64f9676-7724-43e8-b2b6-7036cd86d81a" />

Author: Yes, essentially. spec-gen-sdk orchestrates the same per-language generation scripts that language owners maintain, but runs them in pipeline context (CI) rather than locally. It handles cloning repos, running tsp-client, invoking the language-specific generation scripts, building, and creating PRs.

- [Definitions](#definitions)
- [Background / Problem Statement](#background--problem-statement)
- [Goals](#goals)
- [Workflow Design](#workflow-design)
  - [Workflow Stages](#workflow-stages)
  - [Architecture: Skill Chaining with Next Steps](#architecture-skill-chaining-with-next-steps)
- [Known Gaps](#known-gaps)
- [Success Criteria](#success-criteria)
- [Exceptions and Limitations](#exceptions-and-limitations)
- [Open Questions](#open-questions)

---

## Definitions

- **TypeSpec**: Language for describing cloud service APIs. See [typespec.io](https://typespec.io).
- **SDK**: Client libraries generated from TypeSpec for .NET, Java, JavaScript, Python, Go.
- **Release Plan**: Azure DevOps work item tracking end-to-end SDK release across languages. Managed by azsdk-cli tooling.

> - **API Spec PR**: Pull request in `azure-rest-api-specs` containing TypeSpec changes.

Reviewer 4: ```suggestion
- **API Spec PR**: Pull request in `azure-rest-api-specs` or  `azure-rest-api-specs-pr` containing TypeSpec changes.
```

Author: Done in latest push.

- **APIView**: Web tool for reviewing SDK public API surface.
- **tspconfig.yaml**: Configuration in TypeSpec project specifying emitter settings per language.
- **tsp-location.yaml**: Configuration in SDK repos pointing to the source TypeSpec project.
- **`@azure-tools/typespec-breaking-change`**: TypeSpec-native breaking change detector. Phase A: same-version regression (any diff = error). Phase B: cross-version evolution (request narrowing / response widening = breaking).
- **Suppression Decorators**: `@approvedBreakingChange` (Phase B) and `@approvedUnversionedChange` (Phase A) — inline approval with auto-labeling.
- **TypeSpec Customizations**: SDK-specific customizations in `client.tsp` (renaming, convenience methods).
- **Code Customizations**: Hand-written SDK code that must be preserved across regeneration.

---

## Background / Problem Statement

### Current State

Azure service teams face a complex, multi-step process to release SDKs from TypeSpec specifications:

1. **Fragmented tooling**: Different tools for TypeSpec authoring, spec validation, breaking change detection, SDK generation, and release management — developed independently with limited integration.

> 2. **Manual coordination**: Teams manually track progress across multiple repositories (`azure-rest-api-specs`, `azure-sdk-for-*`) and systems (Azure DevOps, GitHub, Service Tree).

Reviewer 4: This should mention `https://aka.ms/azsdk/releaseplan-dashboard` which is a one-page dashboard to show where a service team is in the process.

Author: Added dashboard link. Thanks!

3. **Silent failures**: SDK generation failures are not surfaced to users — tools fail without actionable guidance.
4. **Scattered documentation**: Process is documented across EngHub, Wiki, and repo specs with no single authoritative source.
5. **No unified CI chain**: Spec PR validation (compile → lint → breaking change → generate → build) lacks a designed end-to-end failure-handling strategy.

### What Has Changed

> - The standalone "release planner" system has been **fully removed**. Release tracking is now handled directly through Azure DevOps work items managed by azsdk-cli.

Reviewer 4: and release plan dashboard `https://aka.ms/azsdk/releaseplan-dashboard`

Author: Added.

- `@azure-tools/typespec-breaking-change` provides TypeSpec-native breaking change detection (replacing OpenAPI-based tools for TypeSpec specs). Crystal's breaking change detection spec defines the CI integration approach.

### Why This Matters

- **Time from API definition to SDK release** is measured in weeks for experienced teams, longer for new ones.

> - **No agent-assisted end-to-end workflow** exists today — each stage requires manual tool invocation and knowledge of the next step.

Reviewer 4: I think this should be modified to include our current agent assisted process. azsdk agent is aware of the current process to some level but not fully.

Author: Updated — now says 'partial agent-assisted workflow exists today via azsdk-cli MCP tools'.

---

## Goals

1. Provide an end-to-end guided workflow from TypeSpec authoring to SDK release
2. Enable resuming the workflow from any intermediate state
3. Automatically track release plan status throughout the workflow
4. Surface errors with actionable guidance at every failure point
5. Integrate sub-skills seamlessly for specialized tasks
6. Support all tier-1 SDK languages with appropriate scope:

| Plane | Required Languages |
|-------|-------------------|
| **Management plane** | .NET, Java, JavaScript, Python, Go |

> | **Data plane** | .NET, Java, JavaScript, Python |

Reviewer 4: Go is optional and some service teams releases Go data plane SDK.

Author: Added Go (optional) to data plane row.

---

## Workflow Design

### Overview

The workflow is an orchestration layer that:
1. **Identifies user intent** and gathers service details
2. **Assesses current state** (existing release plans, PRs, completed steps)
3. **Orchestrates sub-skills** in correct sequence
4. **Tracks progress** via release plan work items
5. **Handles failures** with troubleshooting guidance

### Workflow Stages

The end-to-end flow consists of 5 stages (10 steps). Gaps are annotated inline.

```
                         ┌─────────────────────────────────┐
                         │  ENTRY POINTS                   │
                         ├─────────────────────────────────┤

>                          │ A) User provides initial prompt │

Reviewer 4: In most cases, TypeSpec authoring comes first. Release plan creation requires TypeSpec project path and API release type and a brand-new service team won't even know these details until they prepare a TypeSpec project.

I think we should put TypeSpec authoring as first step and creating a release plan next.

Author: Good call — reordered. TypeSpec authoring is now Step 0, release plan creation is Step 2 (after TypeSpec is ready and path/version are known).

   - Push to `azure-rest-api-specs` and open a PR
   - CI validates automatically:
     - TypeSpec compilation
     - TypeSpec compilation (includes linting)
     - Suppression review (guideline compliance)
     - Breaking change detection (Phase A + B)
     - APIView token generation (for SDK-level review at Stage 4)
     - SDK generation dry-run (spec-gen-sdk)
     - Labels applied based on results

3. 🧑‍💻 **Wait for approvals** *(human gating)*
   - **`PublishToCustomers` label** — required on all PRs targeting `main` or `RPSaaSMaster`. Author self-applies to acknowledge APIs are shipped to customers. Enforced by CI (`summarize-checks`). **Candidate for removal** -- redundant once other gates pass; causes service team confusion.
   - **Namespace approval** — required for first preview of new packages (`namespace-<lang>-pending` → `namespace-<lang>-approved`)
   - **ARM review** — required for ARM/management-plane specs (`ARM-Review-Required` → `ARMSignedOff`)

>    - **Breaking change review** — Spec-level: `BreakingChangeReviewRequired` label auto-applied by CI on any spec PR with breaking changes (review team defined for ARM, undefined for data-plane -- Gap #9). SDK-level: `BreakingChange-{Language}-Sdk` label applies to all specs.

Reviewer 2: `BreakingChange-{Language}-Sdk` labels currently only applies to management plane spec PRs.

Author: updated, throughout

              ┌───────────────────────────────────────────────────┐
              │  STEP 2: Validate & Compile TypeSpec              │
              │  • TypeSpec compiler                              │
              │  • Linter checks                                 │
              │  • Breaking change detection (Phase A + B)       │
              └───────────────────────┬───────────────────────────┘
                                      │
           ┌──────────────────────────┴──────────────────────────┐
           │ FAIL (errors)                                       │ PASS (success)
           ▼                                                     ▼
┌─────────────────────────┐                     ┌─────────────────────────┐
│ Report errors,          │                     │ TypeSpec Ready!         │
│ iterate on TypeSpec     │◀────────────────────│ Extract API version &   │
│ (loop until passing)    │    (if user wants   │ package names           │
└─────────────────────────┘     to fix)         └───────────┬─────────────┘
                                                            │
  Gap: Breaking change findings require manual fix.         ▼

>   No agent auto-resolves suppression decorators.  ┌─────────────────────────┐

Reviewer 4: Namesapce approval comes after a spec PR is created since new namespace approval flow is based on the spec PR itself. right?

Author: Moved — namespace approval now comes after spec PR creation (triggered by the PR itself).

                                                  │ Namespace Approval      │
                                                  │ (new packages only)     │
                                                  │ • Check if pkg name     │
                                                  │   already approved      │
                                                  │ • Request approval if   │
                                                  │   new namespace         │
                                                  └───────────┬─────────────┘
                                                            │
                                                            ▼
              ┌───────────────────────────────────────────────────┐
              │  STEP 3: Open API Spec PR                        │

>               │  (manual — developer opens PR)                   │

Reviewer 4: Agent can also help creating the PR.

Author: Updated Step 3 to say 'agent can help create the PR'.

              │ • TypeSpec compilation                            │
              │ • LintDiff                                        │
              │ • Breaking change detection (Phase A + B)         │
              │ • APIView token generation                        │
              │ • SDK generation dry-run (spec-gen-sdk)           │
              │ • Labels applied                                  │
              └───────────────────────┬───────────────────────────┘
                                      │
           ┌──────────────────────────┴──────────────────────────┐
           │ FAIL                                                │ PASS
           ▼                                                     ▼
┌─────────────────────────┐                     ┌─────────────────────────┐
│ Fix issues, push to PR  │                     │ All checks pass         │
│ (re-triggers pipeline)  │                     │ PR approved & merged    │
└─────────────────────────┘                     └───────────┬─────────────┘
                                                            │
  Gap: Validation steps run independently.                  │

>   No unified PR comment with all results + next steps.      │

Reviewer 4: Can you add ARM review and API review step in this process? E2E experience should outline this offline process that's based on the spec PR.

Author: Added Step 4b: ARM Review & API Review (offline) between CI pass and merge. Shows both ARM review (resource model sign-off) and API review via APIView as blocking steps on the spec PR.

  No designed failure ordering.                             │
                                                            │
                                                            ▼
╔═══════════════════════════════════════════════════════════════════╗
║  STAGE 3: SDK Generation (per language)                          ║
╚═══════════════════════════════════════════════════════════════════╝
                                    │
              ┌───────────────────────────────────────────────────┐
              │  STEP 5: Generate SDKs                           │

>               │  Trigger: spec PR merge → spec-gen-sdk auto-runs │

Reviewer 4: spec-gen-sdk is confusing team. Automated SDK generation creates SDK PR and link it to the release plan.

Author: Replaced 'spec-gen-sdk auto-runs' with 'Automated SDK generation creates SDK PRs & links to release plan' in the diagram. Less confusing for teams unfamiliar with the tool name.

              └───────────────────────┬───────────────────────────┘
                                      │
           ┌──────────────────────────┴──────────────────────────┐
           │                                                     │
           ▼                                                     ▼
┌─────────────────────────┐                     ┌─────────────────────────┐

> │ PIPELINE generation     │                     │ LOCAL generation        │

Reviewer 4: We should make this as automated SDK generation when spec PR merges then update the generated PR when any customization is required. I don't think we need to put local SDK gen in the new flow. Our final goal is to make service team focus only on TypeSpec and SDK generation and release will be a process SDK team automates/handles.

Author: yes, I just realized that and pushed an update to the diagram.

                                      │
           ┌──────────────────────────┴──────────────────────────┐
           │ BUILD FAIL                                          │ BUILD PASS
           ▼                                                     ▼
┌─────────────────────────┐                     ┌─────────────────────────┐
│ Custom code drift?      │                     │ CI green                │
├─────────────────────────┤                     │ Awaiting review         │
│ YES: auto-sdk-build-fix │                     └───────────┬─────────────┘
│ label → Copilot agent   │                                 │
│ auto-repairs → re-runs  │                                 │
├─────────────────────────┤                                 │
│ NO: Pipeline troubleshoot│                                │
│ agent diagnoses          │                                │
└───────────┬─────────────┘                                 │
            │                                               │
            └───────────────────────┬───────────────────────┘
                                    │
                                    ▼
              ┌───────────────────────────────────────────────────┐

>               │  STEP 8: APIView Review                          │

Reviewer 4: This should include new API review process instead of APIView because APIView tool will be retired and switch this to API review.

Author: Should it already be in the spec, because we don't expose any tooling currently around the new system?

│ CI re-runs (loop back   │                                 │
│ to Step 7)              │                                 │
└───────────┬─────────────┘                                 │
            │                                               │
            └───────────────────────┬───────────────────────┘
                                    │
                                    ▼
              ┌───────────────────────────────────────────────────┐
              │  SDK PR approved & merged                         │
              └───────────────────────┬───────────────────────────┘
                                      │
                                      ▼
╔═══════════════════════════════════════════════════════════════════╗
║  STAGE 5: Release Coordination                                   ║
╚═══════════════════════════════════════════════════════════════════╝
                                    │
              ┌───────────────────────────────────────────────────┐
              │  STEP 9: Release SDKs                            │
              │  [Release Skill]                                 │
              └───────────────────────┬───────────────────────────┘

>                                       │

Reviewer 4: Change log update to put release date is missing in this flow.

Author: Added 'Update changelog with release date' as first step in Stage 5.

                                      ▼
              ┌───────────────────────────────────────────────────┐

>               │ • Check release readiness per language            │

Reviewer 4: This should talk about automatic release if a package is ready for auto release.

Author: Added 'Auto-release if package is configured for it' to Stage 5.

| 3 | Steps 5–6: Generate, link PRs | Auto (PRs created by spec-gen-sdk) |
| 4 | Steps 7–8: CI + APIView + repair | Manual (user triggers release) |
| 5 | Step 9: Release | — |

| Stage | Name | Input | Output | Key Tools |
|-------|------|-------|--------|-----------|
| 1 | TypeSpec Authoring | Service requirements | `.tsp` files, `tspconfig.yaml` | TypeSpec compiler, authoring agent, breaking change tool, linter |
| 2 | Spec PR Validation | PR in `azure-rest-api-specs` | Validation checks, labels, APIView tokens | Validation pipeline, LintDiff, `@azure-tools/typespec-breaking-change`, APIView emitter, spec-gen-sdk dry-run |
| 3 | SDK Generation | Approved/merged spec | SDK PRs per language | tsp-client, language emitters, spec-gen-sdk, azsdk-cli (`generate`, `build`, `test`, `metadata`) |
| 4 | SDK PR Validation & API Review | SDK PRs in language repos | Approved & merged PRs | Language CI, SDK breaking change detector, APIView, feedback resolution agent, pipeline troubleshooting agent, auto SDK PR repair |
| 5 | Release Coordination | Merged SDK PRs | Published packages, KPI update | Release plan tooling, release pipelines, Service Tree |

---

#### Stage 1: TypeSpec Authoring

> Developer writes/updates TypeSpec locally: `.tsp` files, `tspconfig.yaml`, local compilation, linting, agent-assisted feedback.

Reviewer 4: TypeSpec authoring should talk about private preview, public preview and stable which depends on product life cycle and where their spec is.

Author: Added API version lifecycle (private preview → public preview → stable) to Stage 1 description. Promotions to stable require breaking change review.

| Tool | Role | Owner |
|------|------|-------|
| Spec PR validation pipeline | Orchestrates full validation suite | EngSys |
| TypeSpec compiler | CI compilation + linting (lint rules run during compile) | TypeSpec team |

> | TypeSpec Suppression Review | Review TypeSpec lint rules that are suppressed. Replacing Swagger-based Spectral LintDiff. Also reviews `suppressions.yaml` changes (manual review, CI reports additions). New label TBD. See [api-reviewer-agent.md](https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/api-reviewer-agent.md). | @catalinaperalta (tool), EngSys (integration) |

Reviewer 6: Small nit: I'm not sure if typespec lint suppressions + suppressions.yaml will be reviewed in the same step.

Author: Updated to note that TypeSpec decorator suppressions and suppressions.yaml changes may be reviewed as separate steps.

| `@azure-tools/typespec-breaking-change` | Same-version + cross-version detection at **TypeSpec/spec level**. Auto-adds `BreakingChangeReviewRequired` / `VersioningReviewRequired` labels. Detects breaking changes in the API spec itself. | @markcowl |
| SDK breaking change detector | Detects breaking changes in the **generated SDK API surface** (complements spec-level detection). Some changes break SDKs but not APIs (e.g., added optional params, parameter order, type renames) and vice versa. Both detectors are needed for complete coverage. | @chunyu3 / @raych1 |
| APIView emitter (`typespec-apiview`) | Generates API surface tokens for SDK-level architect review (tokens used at Stage 4). **Will be retired with ARH.** | APIView team |
| spec-gen-sdk | SDK generation validation (dry-run) | EngSys (@raych1) |
| Avocado / OAV | Legacy Swagger validation -- **being deprecated**, no timeline. Issues overridden when not worth addressing. Exception mechanism being designed for specs that don't emit swagger. | EngSys |

> | typespec-autorest (swagger generation) | Generates swagger JSON from TypeSpec for docs pipeline. **Required for foreseeable future**. README also required while swagger is checked in ([directory structure](https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/directory-structure.md)). | TypeSpec team |

Reviewer 6: I'd say this is partially owned by eng sys as well since the readme requirement is separate from typespec

Author: Added EngSys as co-owner for README enforcement.
